{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "from os import listdir\n",
    "import shutil\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from matplotlib import colors\n",
    "from sklearn.manifold import TSNE\n",
    "import subprocess\n",
    "import shlex\n",
    "import scanpy\n",
    "from sklearn.metrics import r2_score\n",
    "import random\n",
    "\n",
    "# todo:\n",
    "# https://www.embopress.org/doi/full/10.15252/msb.20188746\n",
    "# https://scanpy.readthedocs.io/en/stable/tutorials.html\n",
    "# umap / tsne on the whole dataset to see if ages are naturally seperable\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aging is the time dependant weaking of the body. For example, aging in the pancreous can predispose people to diabetes (https://www.imrpress.com/journal/FBE/10/1/10.2741/e813). I will attempt to find theraputic targets that could be used to target aging in the pancreatus using scRNA-Seq data from  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81547.\n",
    "\n",
    "A good target is a gene or pathway that:\n",
    "\n",
    "(1) Has reletively constant expression accross cells from the same individual. Targeted cell delivery is a hard process, especially between cell types in the same organ, so the genes or pathways need to be consistant accross the single cells.\n",
    "\n",
    "(2) Is highly corralative with age. This is less important because if there is a gene whose expression is dependant on age in some nonlinear maner, it could still be a good target if it is modulated carefully. This is a simplifying assumption.\n",
    "\n",
    "(3) Has a large magnatude of change with age. Something that only changes subtley with age will be harder to modulate. This is more for gene selection rather than pathway selection."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Preprocessing\n",
    "\n",
    "Creates the aggragated data for each sample in the ClupedData folder\n",
    "\n",
    "This just puts the data in a nicer form for downstream processing: files named with the donor age and sex with the counts of each RNA in each row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the description of the data:\n",
    "\n",
    "\tAs organisms age, cells accumulate genetic and epigenetic changes that eventually lead to impaired organ function or catastrophic failure such as cancer. Here we describe a single-cell transcriptome analysis of 2544 human pancreas cells from donors, spanning six decades of life. We find that islet cells from older donors have increased levels of disorder as measured both by noise in the transcriptome and by the number of cells which display inappropriate hormone expression, revealing a transcriptional instability associated with aging. By analyzing the spectrum of somatic mutations in single cells from previously-healthy donors, we find a specific age-dependent mutational signature characterized by C to A and C to G transversions, indicators of oxidative stress, which is absent in single cells from human brain tissue or in a tumor cell line. Cells carrying a high load of such mutations also express higher levels of stress and senescence markers, including FOS, JUN, and the cytoplasmic superoxide dismutase SOD1, markers previously linked to pancreatic diseases with substantial age-dependent risk, such as type 2 diabetes mellitus and adenocarcinoma. Thus, our single-cell approach unveils gene expression changes and somatic mutations acquired in aging human tissue, and identifies molecular pathways induced by these genetic changes that could influence human disease. Also, our results demonstrate the feasibility of using single-cell RNA-seq data from primary cells to derive meaningful insights into the genetic processes that operate on aging human tissue and to determine which molecular mechanisms are coordinated with these processes.\n",
    "\n",
    "This dataset contains scRNA-seq data on the pancreas of humans of various ages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1/2-SBSRNA4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2LD1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23460</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>407558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23461</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>3754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23462</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23463</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23464</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>85563</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23465 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0       1\n",
       "0               1/2-SBSRNA4       0\n",
       "1                      A1BG       0\n",
       "2                  A1BG-AS1       0\n",
       "3                      A1CF       0\n",
       "4                     A2LD1       0\n",
       "...                     ...     ...\n",
       "23460            no_feature  407558\n",
       "23461             ambiguous    3754\n",
       "23462         too_low_aQual       0\n",
       "23463           not_aligned       0\n",
       "23464  alignment_not_unique   85563\n",
       "\n",
       "[23465 rows x 2 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = 'GSE81547'\n",
    "datasets = sorted([path+'/'+i for i in listdir(path)])\n",
    "\n",
    "pd.read_csv(datasets[0],delimiter='\\t',header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The files themselves don't contain any information on the samples. Clumn 0 contains the gene and column 1 contains the Transcript counts. The last 5 rows contains data on the RNA-Seq preprocessing. First lets use the metadata file to get the ages and sexes of each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "id2file = lambda id: [i for i in datasets if i[i.find('/')+1:i.find('_')] == id][0]\n",
    "with open('GSE81547_series_matrix.txt') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('!Series_sample_id'):\n",
    "            identifier = line[len('!Series_sample_id\\t')+1:-3].split(' ')\n",
    "        if line.startswith('!Sample_title'):\n",
    "            donor = [(int(i[1:i.find('yr')]),i[i.find('_')+1]) for i in line[len('!Sample_title\\t'):].split('\\t')]\n",
    "            donor = [(i, [id2file(identifier[j]) for j,k in enumerate(donor) if k == i]) for i in set(donor)]\n",
    "donor.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "for id,files in donor:\n",
    "    dfs = [pd.read_csv(files[0],delimiter='\\t',header=None)] + [pd.read_csv(f,delimiter='\\t',header=None)[1] for f in files[1:]]\n",
    "    totdata = pd.concat(dfs,axis=1).iloc[:-5, :]\n",
    "    totdata.columns = ['Gene'] + list(range(len(totdata.columns)-1))\n",
    "    totdata.to_csv(f'ClumpedData/{\"0\"*(id[0]<10)}{id[0]}_{id[1]}.csv',index=False,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have a function that takes the file name and returns the metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Cleaning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will be using enhance from https://www.biorxiv.org/content/10.1101/655365v2.full\n",
    "\n",
    "The algorithm is incredibly simple. It basically assumes the data has a large nullspace and the data is noisy and uses PCA to choose which dimensions to keep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-02-26 12:36:39] INFO: Reading the expression matrix (24.4 MB) from \"ClumpedData/38_f.csv\"...\n",
      "[2022-02-26 12:36:40] INFO: The expression matrix contains 23460 genes and 474 cells.\n",
      "[2022-02-26 12:36:40] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:36:40] INFO: Input matrix hash: 6252997962fecec3c3afc6b35e100870\n",
      "[2022-02-26 12:36:41] INFO: The median transcript count of the matrix is 559682.5.\n",
      "[2022-02-26 12:36:41] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:36:41] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:36:45] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:36:45] INFO: Aggregating cells...\n",
      "[2022-02-26 12:36:48] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:36:50] INFO: ENHANCE took 9.2 s.\n",
      "[2022-02-26 12:36:50] INFO: Denoised matrix hash: 2cf78b970ff28bbad9f01b66224177b8\n",
      "[2022-02-26 12:36:50] INFO: Writing the denoised expression matrix to \"DenoisedData/38_f.csv\"...\n",
      "[2022-02-26 12:37:01] INFO: File size: 90.1 MB.\n",
      "[2022-02-26 12:37:03] INFO: Reading the expression matrix (14.5 MB) from \"ClumpedData/44_f.csv\"...\n",
      "[2022-02-26 12:37:04] INFO: The expression matrix contains 23460 genes and 277 cells.\n",
      "[2022-02-26 12:37:04] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:37:04] INFO: Input matrix hash: a84467dac2fef1a0ac401d94ca7c3e55\n",
      "[2022-02-26 12:37:04] INFO: The median transcript count of the matrix is 597435.0.\n",
      "[2022-02-26 12:37:04] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:37:04] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:37:06] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:37:06] INFO: Aggregating cells...\n",
      "[2022-02-26 12:37:08] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:37:09] INFO: ENHANCE took 5.2 s.\n",
      "[2022-02-26 12:37:09] INFO: Denoised matrix hash: 93fe26413323f6ab628c3c1711222f72\n",
      "[2022-02-26 12:37:09] INFO: Writing the denoised expression matrix to \"DenoisedData/44_f.csv\"...\n",
      "[2022-02-26 12:37:15] INFO: File size: 50.1 MB.\n",
      "[2022-02-26 12:37:16] INFO: Reading the expression matrix (14.4 MB) from \"ClumpedData/22_m.csv\"...\n",
      "[2022-02-26 12:37:17] INFO: The expression matrix contains 23460 genes and 286 cells.\n",
      "[2022-02-26 12:37:17] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:37:17] INFO: Input matrix hash: b1b7db2b8171c85ef44a6cdac8887947\n",
      "[2022-02-26 12:37:17] INFO: The median transcript count of the matrix is 316511.5.\n",
      "[2022-02-26 12:37:17] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:37:17] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:37:20] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:37:20] INFO: Aggregating cells...\n",
      "[2022-02-26 12:37:22] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:37:23] INFO: ENHANCE took 5.9 s.\n",
      "[2022-02-26 12:37:23] INFO: Denoised matrix hash: 3840a948d003ed9523c291fc67cac437\n",
      "[2022-02-26 12:37:23] INFO: Writing the denoised expression matrix to \"DenoisedData/22_m.csv\"...\n",
      "[2022-02-26 12:37:29] INFO: File size: 50.8 MB.\n",
      "[2022-02-26 12:37:30] INFO: Reading the expression matrix (9.0 MB) from \"ClumpedData/06_m.csv\"...\n",
      "[2022-02-26 12:37:31] INFO: The expression matrix contains 23460 genes and 178 cells.\n",
      "[2022-02-26 12:37:31] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:37:31] INFO: Input matrix hash: 1f26ae1595255383f94fe7084f0114a1\n",
      "[2022-02-26 12:37:31] INFO: The median transcript count of the matrix is 511640.0.\n",
      "[2022-02-26 12:37:31] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:37:31] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:37:33] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:37:33] INFO: Aggregating cells...\n",
      "[2022-02-26 12:37:34] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:37:35] INFO: ENHANCE took 4.2 s.\n",
      "[2022-02-26 12:37:35] INFO: Denoised matrix hash: dc347a781788378232b136706869f4bb\n",
      "[2022-02-26 12:37:35] INFO: Writing the denoised expression matrix to \"DenoisedData/06_m.csv\"...\n",
      "[2022-02-26 12:37:39] INFO: File size: 29.1 MB.\n",
      "[2022-02-26 12:37:40] INFO: Reading the expression matrix (11.1 MB) from \"ClumpedData/01_m.csv\"...\n",
      "[2022-02-26 12:37:40] INFO: The expression matrix contains 23460 genes and 221 cells.\n",
      "[2022-02-26 12:37:40] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:37:40] INFO: Input matrix hash: a49b59c51adb311b17e84c8ed90d7bb6\n",
      "[2022-02-26 12:37:40] INFO: The median transcript count of the matrix is 415386.0.\n",
      "[2022-02-26 12:37:40] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:37:40] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:37:42] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:37:42] INFO: Aggregating cells...\n",
      "[2022-02-26 12:37:44] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:37:45] INFO: ENHANCE took 4.8 s.\n",
      "[2022-02-26 12:37:45] INFO: Denoised matrix hash: 024a5520bc8534faf9adca2349aacaf2\n",
      "[2022-02-26 12:37:45] INFO: Writing the denoised expression matrix to \"DenoisedData/01_m.csv\"...\n",
      "[2022-02-26 12:37:50] INFO: File size: 37.8 MB.\n",
      "[2022-02-26 12:37:51] INFO: Reading the expression matrix (16.8 MB) from \"ClumpedData/05_m.csv\"...\n",
      "[2022-02-26 12:37:52] INFO: The expression matrix contains 23460 genes and 331 cells.\n",
      "[2022-02-26 12:37:52] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:37:52] INFO: Input matrix hash: 89c3879c62bda5dc049da866e7b15f7f\n",
      "[2022-02-26 12:37:52] INFO: The median transcript count of the matrix is 786407.0.\n",
      "[2022-02-26 12:37:52] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:37:52] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:37:55] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:37:55] INFO: Aggregating cells...\n",
      "[2022-02-26 12:37:57] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:37:58] INFO: ENHANCE took 6.6 s.\n",
      "[2022-02-26 12:37:58] INFO: Denoised matrix hash: 5b981c3de895d5f2cf371752833be43a\n",
      "[2022-02-26 12:37:58] INFO: Writing the denoised expression matrix to \"DenoisedData/05_m.csv\"...\n",
      "[2022-02-26 12:38:06] INFO: File size: 60.2 MB.\n",
      "[2022-02-26 12:38:07] INFO: Reading the expression matrix (13.8 MB) from \"ClumpedData/54_m.csv\"...\n",
      "[2022-02-26 12:38:08] INFO: The expression matrix contains 23460 genes and 272 cells.\n",
      "[2022-02-26 12:38:08] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:38:08] INFO: Input matrix hash: 6f3cbeeb31967102874c194f206e7daf\n",
      "[2022-02-26 12:38:08] INFO: The median transcript count of the matrix is 477723.0.\n",
      "[2022-02-26 12:38:08] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:38:08] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:38:10] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:38:10] INFO: Aggregating cells...\n",
      "[2022-02-26 12:38:12] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:38:13] INFO: ENHANCE took 5.5 s.\n",
      "[2022-02-26 12:38:13] INFO: Denoised matrix hash: f0443e8fc3ff92045a2406cd5124a00c\n",
      "[2022-02-26 12:38:13] INFO: Writing the denoised expression matrix to \"DenoisedData/54_m.csv\"...\n",
      "[2022-02-26 12:38:19] INFO: File size: 47.6 MB.\n",
      "[2022-02-26 12:38:20] INFO: Reading the expression matrix (26.3 MB) from \"ClumpedData/21_m.csv\"...\n",
      "[2022-02-26 12:38:21] INFO: The expression matrix contains 23460 genes and 505 cells.\n",
      "[2022-02-26 12:38:21] INFO: Applying ENHANCE...\n",
      "[2022-02-26 12:38:21] INFO: Input matrix hash: 34ae79c3bf2c8321a33bce10644dfce2\n",
      "[2022-02-26 12:38:21] INFO: The median transcript count of the matrix is 581631.0.\n",
      "[2022-02-26 12:38:21] INFO: Will perform denoising with k=1 (value was determined automatically based on a target transcript count of 200000).\n",
      "[2022-02-26 12:38:21] INFO: Determining the number of significant PCs...\n",
      "[2022-02-26 12:38:25] INFO: The number of significant PCs is 50.\n",
      "[2022-02-26 12:38:25] INFO: Aggregating cells...\n",
      "[2022-02-26 12:38:28] INFO: Removing noise using PCA...\n",
      "[2022-02-26 12:38:29] INFO: ENHANCE took 7.8 s.\n",
      "[2022-02-26 12:38:29] INFO: Denoised matrix hash: 9c295c83c85be18e49fc02d7595db5ff\n",
      "[2022-02-26 12:38:29] INFO: Writing the denoised expression matrix to \"DenoisedData/21_m.csv\"...\n",
      "[2022-02-26 12:38:39] INFO: File size: 98.3 MB.\n"
     ]
    }
   ],
   "source": [
    "folder = 'ClumpedData/'\n",
    "for file in listdir(folder):\n",
    "    subprocess.call(shlex.split(f'python3 enhance/enhance.py -f {folder+file} -o DenoisedData/{file} --seed 0'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the data is denoised, I can start removing genes that won't help me find a theraputic target for aging.\n",
    "\n",
    "The first thing to do would be removing genes whose standard deviation is larger within groups that throughout groups. The only reason this is valid is because I already cleaned the data to (hopefully) remove the technical noise. I will also log transform the data, as is standard with scRNA seq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Genes, Cells) = (23460, 221)\n",
      "(Genes, Cells) = (23460, 331)\n",
      "(Genes, Cells) = (23460, 178)\n",
      "(Genes, Cells) = (23460, 505)\n",
      "(Genes, Cells) = (23460, 286)\n",
      "(Genes, Cells) = (23460, 474)\n",
      "(Genes, Cells) = (23460, 277)\n",
      "(Genes, Cells) = (23460, 272)\n",
      "Failed genes due to high in group variance: 16102\n",
      "Possible genes: 7358\n"
     ]
    }
   ],
   "source": [
    "folder = 'DenoisedData/'\n",
    "stds = []\n",
    "means = []\n",
    "for i,f in enumerate(sorted(listdir(folder))):\n",
    "    df = pd.read_csv(folder+f,delimiter='\\t')\n",
    "    counts = df[df.columns[1:]]\n",
    "    genes = df['Gene']\n",
    "    print('(Genes, Cells) =',counts.shape)\n",
    "    genemean = counts.mean(axis=1)\n",
    "    genestd = counts.std(axis=1)\n",
    "\n",
    "    \n",
    "    stds += [genestd]\n",
    "    means += [genemean]\n",
    "\n",
    "stds = np.array(list(stds))\n",
    "means = np.array(list(means))\n",
    "outgroupstd = means.std(axis=0)\n",
    "ingroupstd = np.median(stds,axis=0)\n",
    "print('Failed genes due to high in group variance:',sum(ingroupstd >= outgroupstd))\n",
    "print('Possible genes:',sum(ingroupstd < outgroupstd))\n",
    "potentialgenes = genes[ingroupstd < outgroupstd]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genes Monotonically Corrolated with Age"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's search for any genes that highly corrolate with age. They would be potential targets\n",
    "\n",
    "I use the Spearmans r and p values to find genes that are associated with aging. I am using Spearmanr because I only care that the relashionship is monotonic, not that it is linear."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corrolated genes that also work with the spearmons corrolation criteria: 37\n",
      "AGT\tANKRD26P1\tARVCF\tBAI2\tCCDC65\tCDCA7\tCHRDL2\tCRH\tDGKQ\tDLGAP3\tDLL3\tFIZ1\tGAB3\tHIST1H1D\tJRK\tKIAA1644\tLCN10\tLEF1\tLOC100506134\tLOC389332\tMYBPHL\tORM1\tPCK2\tPID1\tPRKG1\tRBM24\tSAA1\tSHISA7\tSIX3\tSMOC1\tSUGT1P3\tTAF10\tTFF1\tTMEM176B\tTOR2A\tTSIX\tVGF\t"
     ]
    }
   ],
   "source": [
    "# print(means.shape)\n",
    "t = [int(i[:2]) for i in sorted(listdir(folder))]\n",
    "goodinds = []\n",
    "for i,y in enumerate(means[:,potentialgenes.index].T):\n",
    "    if np.max(y) > 1:\n",
    "        r,p = sp.stats.spearmanr(t,y)\n",
    "        if abs(r) > 0.8 and p < 0.05:\n",
    "            goodinds += [potentialgenes.index[i]]\n",
    "\n",
    "\n",
    "print('Corrolated genes that also work with the spearmons corrolation criteria:',len(goodinds))\n",
    "with open('TargetGenesV0.txt','w+') as f:\n",
    "    for i in genes[goodinds]:\n",
    "        print(i,end='\\t')\n",
    "        f.write(i+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reactome Pathway Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aging has many causes, so targeting any specific gene will likely not result in a succesful drug. I will see which pathways are differentially regulated due to age so they could be used as a drug target. The idea is if, for example, immune system pathways are enriched, immunosuprescence could be a potential drug target.\n",
    "\n",
    "\n",
    "I will be using Reactome for my pathway analysis. Having more genes to classify would be more informative, even if it is noisier, so I will be a lot less stringant on deterining if a gene is related with age. In order to accomplish this, I will allow for some nonlinearity in the data (using parabolic regress instead of linear) and have a reletively lower for being corrolated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reactome genes with a good parabolic fit: 30\n"
     ]
    }
   ],
   "source": [
    "# print(means.shape)\n",
    "t = [int(i[:2]) for i in sorted(listdir(folder))]\n",
    "goodinds = []\n",
    "for i,y in enumerate(means[:,potentialgenes.index].T):\n",
    "    if np.max(y) > 1:\n",
    "        poly = np.poly1d(np.polyfit(t, y, 2))\n",
    "        yhat = poly(t)\n",
    "        r = r2_score(y,yhat)\n",
    "        if r > 0.7:\n",
    "            goodinds += [potentialgenes.index[i]]\n",
    "\n",
    "\n",
    "print('Reactome genes with a good parabolic fit:',len(goodinds))\n",
    "with open('ReactomeGenesV0.txt','w+') as f:\n",
    "    for i in genes[goodinds]:\n",
    "        f.write(i+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check out some of these genes to make sure they're actually related to age... (Just run the cell over and over for different examples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Gene Expression')"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXgAAAEWCAYAAABsY4yMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAnHElEQVR4nO3dd5wV5fXH8c9Zli4oVSkColiwoayIYsUSOyoYxYYRxBhNNIkaNdFoYvylR5MgVqzYEEFEsYIiRkCWiAUVFEWqFJEusLvn98dzNy645cLu7Nw79/t+ve5rd+aWOePK2dkzz3Mec3dERCR58uIOQEREoqEELyKSUErwIiIJpQQvIpJQSvAiIgmlBC8iklBK8CIiCaUEL5Ezsy/MbL2ZrSnzaJt6bhczKzGzO8t5n5vZbqnvb05t/2yL11yV2n9zarunmb1iZl+b2VIzG2FmbWrhNKvNzI4xs4/NbJ2ZTTCzjpW8ds0Wj2Iz+1eZ5xuZ2Z1mtszMVprZxNo5C8kkSvBSW0519+3KPBam9l8IrADOMbP6VXzGLGDAFvsuTO0v1Qy4B+gEdARWAw9UN/jymFl+DX5WS+AZ4EagOTANeLKi15f9bwnsCKwHRpR5yT2pz9kr9fXnNRWrZA8leInbhcBvgE3AqVW89h2gkZntDZD62jC1HwB3H+fuI9x9lbuvA/4N9KqpYFN/LVxuZrOB2TX1ucCZwIep2L8Fbgb2N7M903hvP2AJ8GYqxj2A04DB7r7U3YvdvbAGY5UsoQQvsTGzw4H2wBPAU4RkX5VHyrxuAPBwFa8/AviwzDHPNbP3tj7azZwOHAx0Le9JM/umksd1FXzm3sCM0g13Xwt8ltpflQHAw/5d35GDgbnALakSzftm1je9U5MkUYKX2jK6TJIbndo3ABjn7iuAx4ATzax1FZ/zKNDfzOoC56S2y2Vm+wE3AdeU7nP3x9x9v2qcB8D/ufvX7r6+vCfdfYdKHn+s4DO3A1ZusW8l0KSyQMysA3Ak8FCZ3e2BfVLvbwtcATxkZntVfWqSJErwUltOL5PkTjezhsBZwHAAd38b+BI4t7IPcfcvgU+B24DZ7j6vvNelbs6OA6509zdr8DwAyj1mNa0Bmm6xrynhHkJlLgQmufvnZfatJ5S8bnX3je7+BjABOL6mgpXsoAQvcTmDkMDuNLPFZrYYaEd6ZZqHgV9SQXkmNfrkVeD37v5IDcVbVqUtWMsZ4VL2cUMFb/sQ2L/MZzQGdqVMeakCF7L51TtAdUtQkhBK8BKXAcAwYF+gW+rRC+hmZvtW8d4nCVejT235hJm1A8YDQ9z9rhqMN21bjBba8nFbBW8bBexjZn3NrAGhtPSeu39c0XHM7FDCL8URWzw1kfDX0PVmlm9mvYCjgJeqe26SXZTgpdalkvAxwO3uvrjMoxB4ke8PhdyMu69391crqIEPAjoDvy175Vzm2OeZWVVXxbXO3ZcCfYE/EIaNHky4xwCAmd1gZuO2eNsA4Bl336yM4+6bgD7ASYQ6/L3AhZX9spBkMi34ISKSTLqCFxFJKCV4EZGEUoIXEUkoJXgRkYSqsWZJNaFly5beqVOnuMMQEckahYWFy9y9VXnPZVSC79SpE9OmTYs7DBGRrGFmcyt6TiUaEZGEUoIXEUkoJXgRkYRSghcRSSgleBGRhFKCFxFJKCV4EZEYFc5dwZAJn1I4d0WNf3ZGjYMXEcklhXNXcN59k9lYVEK9/DyGD+pJ947NauzzdQUvIhKTyXOWs7GohBKHTUUlTJ6zvEY/XwleRCQmPTu3oF5+HnUM6ubn0bNzixr9fJVoRERi0r1jM4YP6snkOcvp2blFjZZnQAleRCRW3Ts2q/HEXkolGhGRhFKCFxFJKCV4EZGEUoIXEUkoJXgRkYSKNMGb2Q5m9rSZfWxmH5nZIVEeT0REvhP1MMk7gBfdvZ+Z1QMaRXw8ERFJiSzBm1lT4AjgIgB33whsjOp4IiKyuShLNJ2BpcADZvZfM7vPzBpv+SIzG2xm08xs2tKlSyMMR0Qkt0SZ4POBA4Gh7n4AsBa4bssXufs97l7g7gWtWrWKMBwRkdwSZYKfD8x39ymp7acJCV9ERGpBZAne3RcD88xsj9SuY4CZUR1PREQ2F/Uomp8Cw1MjaOYAP4r4eCIikhJpgnf3d4GCKI8hIiLl00xWEZGEUoIXEUkoJXgRkYRSghcRSSgleBGRhFKCFxFJKCV4EZGEUoIXEUkoJXgRkYRSghcRSSgleBGRhFKCFxFJKCV4EZGEUoIXEUkoJXgRkTjNegkm/hWKNtb4RyvBi4jEZdN6eOFqeH8EmNX4x0e9opOIiFRk0j/gmy9hwFioU7fGP15X8CIicfh6Dky6HfbpB7scHskhlOBFROLw4vXhqv34WyM7hBK8iEht+2QczHoRjroOmraJ7DCR1uDN7AtgNVAMFLm7FuAWkdy2cR2MuxZa7QkH/zjSQ9XGTdaj3X1ZLRxHRCTzTfp7uLF60fOR3FgtSyUaEZHasmw2vHUH7Hc2dDos8sNFneAdeNnMCs1scMTHEhHJXO5hzHt+Qzju97VyyKhLNL3cfaGZtQZeMbOP3X1i2RekEv9ggA4dOkQcjohITD58Bua8Dif9FZrsWCuHjPQK3t0Xpr4uAUYBPcp5zT3uXuDuBa1atYoyHBGReHy7El68AdrsDwUX19phI0vwZtbYzJqUfg8cD3wQ1fFERDLW+FthzVdwyj8gr06tHTbKEs2OwCgL/RXygcfc/cUIjyciknkWTIep98JBg6Bd91o9dGQJ3t3nAPtH9fkiIhmvuAjGXgXbtYZjbqz1w6vZmIhIVN65DxbNgH7DoMH2tX54jYMXEYnCygUw/vew6zGw95mxhKAELyIShXHXQkkxnPy3SHq9p0MJXkSkpn00Fj4eC0f9CprvElsYSvAiIjXp21XwwjWw4z5wyBWxhqKbrCIiNWn8rbB6EZz9SOTNxKqiK3gRkZoy7x2Yeg/0uATax98dXQleRKQmFG2E534GTdvCMTfFHQ2gEo2ISM146w5YMhP6Pwn1m8QdDaAreBGR6ls6Cyb+OYx33+OEuKP5HyV4EZHqKCkJpZm6jeDEP8UdzWZUohERqY5p98OXb0OfIaHnTAbRFbyIyLb65kt49WbYtTd0Oy/uaL6nyit4M9sduAboWPb17t47wrhERDKbO4z9efh6yu2xtSOoTDolmhHAXcC9QHG04YiIZIkZT8Cnr8KJf4ZmHeOOplzpJPgidx8aeSQiItli9Vfw4nWwc0846JK4o6lQOjX458zsJ2bWxsyalz4ij0xEJBO5w/O/gE3roc+/IS9zb2WmcwU/IPX1mjL7HOhc8+GIiGS4D0aGTpHH/Q5adok7mkpVmeDdPb5elyIimWTN0tApsl332DtFpiOdUTR1gcuAI1K7XgfudvdNEcYlIpJ5XvglbFwDfe6EvDpxR1OldEo0Q4G6wJ2p7QtS+wZFFZSISMb54BmY+WxoJNZ6z7ijSUs6Cf4gd9+/zPZ4M5uR7gHMrA4wDVjg7qdsbYAiIrFbswSe/yW0PRAOvTLuaNKWzu3fYjPbtXTDzDqzdePhrwQ+2trAREQyQumEpo1r4fShUCd7OrykE+k1wAQzmwMYYUbrj9L5cDNrD5wM/AH4xbYGKSISm/dHfDdqJktKM6XSGUXzmpl1AfYgJPiP3X1Dmp9/O3AtUGFzZDMbDAwG6NChQ5ofKyJSC1YtCqNm2vfIilEzW6qwRGNmvVNfzyRche8G7AqcnNpXKTM7BVji7oWVvc7d73H3AncvaNWq1VYFLyISGXcYcwUUbQilmSwYNbOlyq7gjwTGA6eW85wDz1Tx2b2A08zsJKAB0NTMHnX387cpUhGR2lT4QKrXzF+g5W5xR7NNzN2jP4jZUcDVVY2iKSgo8GnTpkUej4hIpb6eA0MPg50PgvNHZXQ7AjMrdPdyV/iuMmozu9LMmlpwn5lNN7Pjaz5MEZEMUFIMoy6DvPywiEcGJ/eqpBP5xe6+CjgeaE0YQfPHrTmIu7+uMfAikhXeugPmTYaT/gzbt487mmpJJ8GXdrE/CXjA3WeU2ScikhyLZsCE26BrH9jv7LijqbZ0Enyhmb1MSPAvmVkToCTasEREatmm9TDyEmjUImNXaNpa6Ux0Ggh0A+a4+7pUL/i0JjqJiGSNV2+BZZ/A+SOhUTKWvEjnCv4Q4BN3/8bMzgd+A6yMNiwRkVr02XiYMhR6DIbdjo07mhqTToIfCqwzs/0Js1LnAg9HGpWISG1ZuzyMmmm5Bxx7S9zR1Kh0EnyRh8HyfYA73P0OKmk9ICKSNdxhzE9h/dfQ9z6o1yjuiGpUOjX41WZ2PaEP/OGp9r91ow1LRKQWTH8IPnkejv8DtNkv7mhqXDpX8GcDGwjj4RcD7YC/RBqViEjUls2GF6+HzkdBz5/EHU0kqkzwqaQ+Eqif2rUMGBVlUCIikSraAE//CPIbwOl3ZfVs1cqk06rgEuBp4O7UrnbA6AhjEhGJ1qu3wOL34fQ7oWmbuKOJTDq/ti4ndIZcBeDuswktC0REss+sl2HykDAkco8T444mUukk+A3uvrF0w8zyCe2CRUSyy+rFMPoyaL03HPf7uKOJXDoJ/g0zuwFoaGbHASOA56INS0SkhpUUwzODw9qq/YZB3QZxRxS5dBL8r4ClwPvApcALhNmsIiLZY9Lf4fM3QpfILFtbdVtVOg7ezPKA99x9H+De2glJRKSGzf1P6BK5Tz844IK4o6k1lV7Bu3sJMMPMtBq2iGSndV/DyEHQrBOc8o9EdIlMVzozWdsAH5rZVGBt6U53Py2yqEREakJJCYz6MaxdCgNfgQZN446oVqWT4JPVfUdEcsd//gmzXwoLZ7ftFnc0ta7KBO/ub5jZTkAPwvDId1KzW0VEMtfct+G130HX06HHJXFHE4t0ZrIOAqYCZwL9gMlmdnHUgYmIbLO1y+Dpi6FZRzjtXzlVdy8rnRLNNcAB7r4cwMxaAP8BhlX2JjNrAEwk9LDJB552999WL1wRkSqUjndftxwGvZpzdfey0knw84HVZbZXA/PSeN8GoLe7rzGzusAkMxvn7pO3IU4RkfRM/At89hqcekciWwBvjXQS/AJgipk9S6jB9wGmmtkvANz97+W9KbVIyJrUZt3UQy0ORCQ6n74Kr/8R9j8XDhwQdzSxSyfBf5Z6lHo29bXKVZ1Si4MUArsBQ9x9SjmvGQwMBujQQcPtRWQbfTMvjHffcW84+W85W3cvK50E/yd3/7bsDjNr6e7LqnqjuxcD3cxsB2CUme3j7h9s8Zp7gHsACgoKdIUvIluvaAOMGBDq7z98OHFL722rdHrRTDWznqUbZtaXcJM1be7+DfA6cMLWvE9EJC3jroUFhaG/e4td444mY6RzBX8eMMzMXgfaAi2A3lW9ycxaAZvc/RszawgcC/ypGrGKiHzf9Ieh8EE47Oew16lxR5NR0pno9L6Z/QF4hDCC5gh3n5/GZ7cBHkrV4fOAp9x9bLWiFREpa8F0eP7qsK5q7xvjjibjVJngzex+YFdgP2B34Dkz+7e7D6nsfe7+HnBAjUQpIrKlNUvhyQtgu9bQdxjk1Yk7ooyTTg3+A+Bod//c3V8CegIHRhuWiEglijfBiItg3TI4+xFo3CLuiDJShQnezJoCuPs/UmPaSW2vRA3IRCROL/0a5k6CU/8JbVUoqEhlV/Cvl35jZq9t8dzoKIIREanSf4fD1Luh509g/7PjjiajVZbgy84SaF7JcyIitWPeOzD2Kuh0eE4sml1dlSV4r+D78rZFRKK1aiE8eR40aQNnPQR10hnlndsq+y/UOtVvxsp8T2q7VeSRiYiU2rQenjgPNq6FC0brpmqaKkvw9/Jdv5my3wPcF1lEIiJlucOYn8HC6XDOY7Bj17gjyhoVJnh310gZEYnfpL/D+09B79/AnifHHU1WSWccvIhIPGaOCcvu7XsWHH513NFkHSV4EclMC9+FUZdC+4PgtH+r/e82UIIXkcyzaiE83h8aNg9197oN4o4oK6Wz6PaOZna/mY1LbXc1s4HRhyYiOWnDGnjsbNiwCs59IvSakW2SzhX8g8BLhFbBALOAqyKKR0RyWUlxWJXpqw/grAdhp33jjiirpZPgW7r7U0AJgLsXAcWRRiUiuemlX8OscXDin6HLcXFHk/XSSfBrzawFqdmrqdWdVkYalYjknslDYcrQ0GOmxyVxR5MI6cz1/QUwBtjVzN4izGLtF2lUIpJbZo6BF68PKzIdf2vc0SRGOis6TTezI4E9CG0KPnH3TZFHJiK5Yd5UeOYSaF8AZ96rhTtqULrdenoAnVKvP9DMcPeHI4tKRHLDsk/h8XOgaVvo/wTUbRh3RImSzpJ9jxCW7HuX726uOqAELyLbbvVX8OgZgMF5T0PjlnFHlDjpXMEXAF3LruokIlIt366C4X1h7XK46DlosWvcESVSumuy7rS1H2xmO5vZBDP7yMw+NLMrtz48EUmcog3w5Pmw5CP44cPQrnvcESVWOlfwLYGZZjYV2FC6091Pq+J9RcAvUzdpmwCFZvaKu8/c9nBFJKuVFMMzg+HzN+D0u6DLsXFHlGjpJPibt+WD3X0RsCj1/Woz+whoByjBi+Qid3jhapg5OgyF7NY/7ogSL51hkm+YWUegi7u/amaNgK0ax2RmnYADgCnbFKWIZL8Jt8G0YdDrKjj0p3FHkxPSaTZ2CfA0cHdqVztgdLoHMLPtgJHAVe6+qpznB5vZNDObtnTp0nQ/VkSyyeShMPHPcMAFcOzNcUeTM9K5yXo50AtYBeDus4G02ruZWV1Cch/u7s+U9xp3v8fdC9y9oFUrLfUqkjjTH4EXrwuzVE+5XX3da1E6CX6Du28s3TCzfFJ9aSpjZgbcD3zk7n/f9hBFJGt9OAqe+xnsegz0vR/qpDu3UmpCOgn+DTO7AWhoZscBI4Dn0nhfL+ACoLeZvZt6nFSNWEUkm8x6GUZeAjsfDGc/Cvn1444o56Tz6/Q6YCDwPnAp8AJwX1VvcvdJhN41IpJrPpsQxrrvuDec+yTUaxR3RDkpnVE0JcC9qYeISOW+eCsst9diN7hgFDTYPu6IclaFJRoz62Nml5fZnmJmc1KPs2onPBHJKvPegcd+CDvsDBc+C42axx1RTqusBn8toQ98qfrAQcBRwI8jjElEstH8Qnj0TGjcCi4cA9tpVFzcKivR1HP3eWW2J7n7cmC5mTWOOC4RySYLCuGRM8IV+0XPQ9M2cUckVH4F36zshrtfUWZTv5pFJFgwPST3hjvAgLGwfbu4I5KUyhL8lNQs1s2Y2aXA1OhCEpE4Fc5dwZAJn1I4d0XVL54/DR4+HepvDxeNDbV3yRiVlWh+Dow2s3OB6al93Qm1+NMjjktEYlA4dwXn3TeZjUUl1MvPY/ignnTv2Kz8F385BR7tC41bwIDnYIcOtRusVKnCBO/uS4BDzaw3sHdq9/PuPr5WIhORWjd5znI2FpVQ4rCpqITJc5aXn+Dn/geGnwXbtVZZJoOlMw5+PKCkLpIDenZuQb38PDYVlVA3P4+enVt8/0WfjYfHz4Xt24crd91QzVhqDCEi/9O9YzOGD+rJ5DnL6dm5xfev3j8ZB09dCC26wIWjwxW8ZCwleBHZTPeOzcovy3wwMqzGtNN+cP5ITWLKAuk0G8t87z0FK+fHHYVIck0bBk8PhPYHaYZqFsn+BL/u67AM2P0/gKWz4o5GJHne/DuM/Tl0OQ7OfwYaNI07IklT9if4Rs3DjZ7iDfDACWHShYhUX0kJvHwjvHYL7NMPznlMXSGzTPYneKBwYwce7XoPG/IawkOnhlalIrLtijfB6MvgP/+EgwbBmfdCnbpxRyVbKesTfOnEjJsmree4lTewrnH7MD73vRFxhyaSnTasgcfPgfeegKN/DSf9FfKyPlXkpKz/qZWdmLGgaAeGd70LOvSEZwbBW3eAV7m6oIiUWrM09VfweDj1DjjyWq2hmsWyPsGXTsyoY1A3P48Dd+8UhnDtfQa8clNY7LekOO4wRTLfstlw3zGw5KOwxF73i+KOSKop68fBVzgxo+8waNIWJg+Bb+ZB33uhnroci5Try8mhLGN1QtOw9gVxRyQ1IOsTPFQwMSMvD064DZp1ghd/BQ+eDP2fhCY7xhKjSMZ6/2kY/ZPQeuD8p6F557gjkhqS9SWaKh08OAzvWvpJ+PNz8QdxRySSGdzh9T/ByIHhin3Qq0ruCRNZgjezYWa2xMziz6h7nAg/GgclRTDsB6Gfhkgu2/QtjLoUXr8N9j83LI6t2amJE+UV/IPACRF+/tZp2w0umRBWen+8P7z1T42wkdy0enEoWb73JPT+DZx+J+TXjzsqiUBkCd7dJwJfR/X526Rpm3Al3/U0eOVGGPVj2LQ+7qhEas+C6XDP0bBkJvzwETjiGg2DTLDYa/BmNtjMppnZtKVLl0Z/wHqNoN+DYQLHe0/AAyfBqoXRH1ckbjOehAdOhLw6MPDlcKEjiRZ7gnf3e9y9wN0LWrWqpbW88/LCBI5zHoNls+DuI8MKNSJJVLwJXrweRg2GdgWhVLnTvnFHJbUg9gQfqz1PDiMH6jcJs/cmD1VdXpJlzRJ45AyYfCccfFlqkY5aupCS2OV2ggdovRcMngBdfhBmvY4cFHpxiGS7uW/DXYfD/HfgjLvhxD+qYViOiXKY5OPA28AeZjbfzAZGdaxqa7B9mJrd+0b48Bm4tzcs+TjuqES2jTu8PSSMlKnXCAa9BvufE3dUEoPIZrK6e/+oPjsSeXlwxNVhxZqRA+Heo+GUf+gfhmSX9Svg2Svg47Gw5ylhCGSD7eOOSmKiEs2WOh8Jl74JbQ8IE0FGXw4b18YdlUjV5r0Ddx0Bs16CH9wW/ipVcs9pSvDladoGLhwDR1wL7w4Po2wWvx93VCLlKymBSf8IK5oZcPFLcMjlGt8uSvAVqpMPvX8NA8bAhtVw7zHw9p3hH5NIpli1EB7pA6/eHEaFXfomtO8ed1SSIZTgq7LLEXDZW7Brb3jpenj0TFi1KO6oRGDmGBh6KMwvhD5D4KyHoOEOcUclGUQJPh2NW0L/x8NN1y8nw9BD4MNRcUcluerblaHNxlMXhHbYl06EA85XSUa+Rwm+CoVzVzBkwqcUfvkNFFwMP34z/KMacRE8fTGsy6x2O0n0v5/B3BVxhxK/Oa/D0F7w3lNw5K9g4CvQcre4o5IMlYgFP6JSuqD3xqIS6uXnMXxQT7p37AIDXw03td74I3wxCU65HfY8Ke5wE6n8n0Gzqt+YNBtWw8s3QuEDoSPqwFdUa5cq6Qq+EmUX9N5UVMLkOcvDE3Xy4chr4JLx0KglPNEfnh4Ia5fFG3ACVfgzyCWfjYc7D4HCB+HQn8KPJym5S1qU4Cux5YLePTu32PwFbfaHwa/DUTfAzGdhSI/wp7P62dSYKn8GSbZ2OTxzaeglk98gdIA8/lao2zDuyCRLmGdQMiooKPBp06bFHcZmCueu+P6C3uXt/2omjLkCFhSGETcn/03Ln9WQin4GieUeLhReuj7cUD3s53D41VC3QdyRSQYys0J3L3eVdCX4bVBhXbikGN65H177HZRsCq0PDv2ZVsuR9C2dBc//Ar54M7T2Pe2fsOPecUclGayyBK8SzTaosC6cVycs8n3FVOhyPIy/NdROP3013oAl821YEyYrDT0UFr8XhuQOfEXJXapFCX4bVFkXbtoWzn4Ezh8Zth/tC4+fC8s/q/1gJbOVlmP+XRBGZu17FlxRGIbk5umfp1SPSjTbKO26cNEGePvfMPFvoWzT87JQT23QtPaClcw07x14+dcwb0pobnfiX2Dng+KOSrKMavCZYNWiUJuf8VgYWnnUddD9Ii3AkItWzIXXboEPRsJ2O0Lv30C383XFXo6cu8G+DZTgM8mC6fDKTeEmWvNd4ZiboGsfTTPPBWuXwcS/wrT7weqEMe29roT628UdWUbSJLf06CZrJml3IAx4Dvo/Ga7eRwyAe46C2a9q/HxSfbsSJvwf3LE/TL07LCLz08LQrVTJvUKa5FZ9alUQBzPY4wTocly4wfb6bTC8L+zcM5RuOh+lK/ok+HYVTLkb3v5XSPJ7nQq9b4JWu8cdWVYoHcywqagk9ya51RCVaDJB0UaY/lAYRbFqAex8cFhsZLdjlOiz0foVIbFPHgrffgO7nxh+cbftFndkWUc1+KqpBp8tijbAfx+BN/8Bq+bDTvuGWYxdTw9j7HNAVv+DXrUIptwVJrttXB0S+5HXQDv1jZHoxJbgzewE4A6gDnCfu/+xstfnfIIvVbQR3n8KJt0Oy2eH9sQHXwYHnAf1m8QdXWSy9qbako/gP/+G954ELw6/kA//Jey0T9yRSQ6oLMFHVoM3szrAEOA4YD7wjpmNcfeZUR0zMfLrhQUc9j8XPh4bxtG/+CuYcBsceAEcNAia7xJ3lDWuvJtqGZvgS4rD4tZT7oLP34C6jaDgR2Geg3oQSYaI8iZrD+BTd58DYGZPAH0AJfh05eVB19PCY/60kOgnD4W3h4QbtAcNgt2OTUz5Jituqq3+Ct59FAofgm/mQtN2Yahr9x9Bo+ZxRyeymchKNGbWDzjB3Qelti8ADnb3K7Z43WBgMECHDh26z507N5J4EmPVwtAXvPBBWPNVSDDdzgvlm2adYg6u+jKyBl9cBHMmhPsjHz8PJUXQ8TDocQnseUpYH0AkJrHU4M3sLOAHWyT4Hu7+04reoxr8VijaCLPGwfSH4dPXAIeOvWC/s2Hv06HB9nFHmN3c4asPw72QGU/CmsXQsDl0OzfMQG7ZJe4IRYCYavCEuvvOZbbbAwsjPF5uya8XZsB27QPfzIMZj8OMJ+C5n8EL14QSzt5nwB4nQr3GcUebPZZ9CjNHwftPw9KPw4zT3X8QEnuXH4T/7iJZIsor+HxgFnAMsAB4BzjX3T+s6D26gq8m97DgyHtPhRWm1iyG/IZhAZI9T4bdT4DGGVjXjlNJCSyeAZ+Mg5ljYOlHYX+HQ2HfvmFETOOWsYYoUplYruDdvcjMrgBeIgyTHFZZcpcaYAbtC8LjhP+DLyfDzNHw8QvwyfNgeWERiS7HhZuzbbrlZoOr9Svg84mhT/+sl8MvQssLSf2EP8Fep8D27eOOUqTaNNEpF7jDohnwyQsw+xVY+F/AoWEz6HQY7HIkdDwUWu2VzIT/7SqYNxXmvhWGNC78L3gJ1GsSZgvvnmoboSt1yUKaySqbW7sMPhsfkt2cibDyy7C//vawcw9of1DoT97uwOxLeiXFsGw2LJwO898Jw0u/+iAk9Lx8aHsg7Ho0dD46/KWjds2S5ZTgpWLusOKLsOjEl2/Dl1PCzUVS/180bQetu4al41rvBS26QMvd4h+lU1IS+vYsmwVLPwkxf/VBWPy8aH14Tf2m4ZdU+x7QqVf4xaUbzpIwcY2ikWxgFmbFNt8ltLEF2LA6lHQWTE8lzQ9hzuthRapSjVrCDh2gWcdQr95uJ2iyU1jAolHzUP5psAPUbbh1DdNKimHDqlBWWbcs/LWxZgmsXhQS+soF4RfSN3OheON372vYPPwSKrgY2uwX7i+03D2ZJSeRNCnBy/fVbxJq850O+25f0caQWJfPDlfNK74IKxMtei/cxC3eUP5nWR7U2y5M5a9TL0wKyqsLeCiblBSHRF30LWz6FjatrTiuRi3Deret9wzDP5vvEv6iaLUnbNeqBv8DiCSDErykJ79e6GPeanfg5M2fcw9tcVd/BWuXwLqvYd3ysG/jOti4NiTu4qKQzEs2hcRf+sivD/kNwqN+k1BaadAUGrWAxq3C1yZtoG6DGE5cJHspwUv1mYWSTMNmwJ5xRyMiKSpQiogklBK8iEhCKcGLiCSUEryISEIpwYuIJJQSvIhIQinBi4gklBK8iEhCZVSzMTNbClS1KGtLYFkthBOnXDhHyI3zzIVzhNw4z0w9x47uXm6vjoxK8Okws2kVdU5Lilw4R8iN88yFc4TcOM9sPEeVaEREEkoJXkQkobIxwd8TdwC1IBfOEXLjPHPhHCE3zjPrzjHravAiIpKebLyCFxGRNCjBi4gkVNYkeDM7wcw+MbNPzey6uOOpKWY2zMyWmNkHZfY1N7NXzGx26muzOGOsLjPb2cwmmNlHZvahmV2Z2p+082xgZlPNbEbqPG9J7U/UeQKYWR0z+6+ZjU1tJ/EcvzCz983sXTObltqXVeeZFQnezOoAQ4ATga5AfzPrGm9UNeZB4IQt9l0HvObuXYDXUtvZrAj4pbvvBfQELk/9/JJ2nhuA3u6+P9ANOMHMepK88wS4EviozHYSzxHgaHfvVmb8e1adZ1YkeKAH8Km7z3H3jcATQJ+YY6oR7j4R+HqL3X2Ah1LfPwScXpsx1TR3X+Tu01PfryYkhnYk7zzd3dekNuumHk7CztPM2hMW5r2vzO5EnWMlsuo8syXBtwPmldmen9qXVDu6+yIIyRFoHXM8NcbMOgEHAFNI4HmmShfvAkuAV9w9ied5O3AtUFJmX9LOEcIv55fNrNDMBqf2ZdV5Zsui21bOPo3vzDJmth0wErjK3VeZlfdjzW7uXgx0M7MdgFFmtk/MIdUoMzsFWOLuhWZ2VMzhRK2Xuy80s9bAK2b2cdwBba1suYKfD+xcZrs9sDCmWGrDV2bWBiD1dUnM8VSbmdUlJPfh7v5ManfizrOUu38DvE64v5Kk8+wFnGZmXxBKpb3N7FGSdY4AuPvC1NclwChCqTirzjNbEvw7QBcz28XM6gHnAGNijilKY4ABqe8HAM/GGEu1WbhUvx/4yN3/XuappJ1nq9SVO2bWEDgW+JgEnae7X+/u7d29E+Hf4Xh3P58EnSOAmTU2syal3wPHAx+QZeeZNTNZzewkQu2vDjDM3f8Qb0Q1w8weB44itCL9CvgtMBp4CugAfAmc5e5b3ojNGmZ2GPAm8D7f1W1vINThk3Se+xFuvNUhXDw95e6/M7MWJOg8S6VKNFe7+ylJO0cz60y4aodQyn7M3f+QbeeZNQleRES2TraUaEREZCspwYuIJJQSvIhIQinBi4gklBK8iEhCKcGLAGZ2hpm5me0ZdywiNUUJXiToD0wiTN4RSQQleMl5qR45vYCBpBK8meWZ2Z2pvu5jzewFM+uXeq67mb2RakL1UunUdZFMowQvElq+vujus4CvzexA4EygE7AvMAg4BP7XU+dfQD937w4MAxIxq1qSJ1u6SYpEqT+hDQaEBlr9Cb3cR7h7CbDYzCaknt8D2IfQXRBCW4JFtRqtSJqU4CWnpXqL9Ab2MTMnJGznuz4k33sL8KG7H1JLIYpsM5VoJNf1Ax52947u3snddwY+B5YBfVO1+B0JDeEAPgFamdn/SjZmtnccgYtURQlecl1/vn+1PhJoS1iH4APgbkLny5WpJSP7AX8ysxnAu8ChtRatyFZQN0mRCpjZdu6+JlXGmUpY4Wdx3HGJpEs1eJGKjU0t4FEP+L2Su2QbXcGLiCSUavAiIgmlBC8iklBK8CIiCaUELyKSUErwIiIJ9f8WduoE8sG6vwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "thay = np.linspace(min(t),max(t),100)\n",
    "g = random.choice(goodinds)\n",
    "y = means[:,g].T\n",
    "poly = np.poly1d(np.polyfit(t, y, 2))\n",
    "yhat = poly(t)\n",
    "r = r2_score(y,yhat)\n",
    "plt.plot(t,y,'.')\n",
    "plt.plot(thay,poly(thay))\n",
    "plt.title(f\"{genes[g]}:   r = {round(r,2)}\")\n",
    "plt.xlabel('Age')\n",
    "plt.ylabel('Gene Expression')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These genes were then ran through Reactome for pathway analysis.\n",
    "\n",
    "Section 4 shows the most upregulated pathways. There's plenty of pathways for a biologist to interperate. The most important thing to look at is the False Discovery Rate (FDR). It can be thought of as the probability that the pathway is differentially expressed through chance alone, as opposed to through biologically relevent meanings.\n",
    "\n",
    "For a visual interprotation, see the image below.\n",
    "\n",
    "The more yellow a pathway is, the more confident Reactome is that relates to aging. For example, genes associated with Neutrophil degranulation are significantly associated with aging in the pancreous.\n",
    "\n",
    "![alt text](ReacfoamV0.jpg \"PathwaysV0\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlap\n",
    "\n",
    "This section does not depend on variables from previous sections"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's see which of these pathways could be targeted using the gene list we found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('ReactomeDataV0.csv')\n",
    "genes = set(pd.read_csv('TargetGenesV0.txt',header=None)[0])\n",
    "\n",
    "gene2path = { i : [] for i in genes}\n",
    "for i , j , fdr in zip(data['Submitted entities found'],data['Pathway name'],data['Entities FDR']):\n",
    "    if fdr < 0.05:\n",
    "        for k in i.strip().split(';'):\n",
    "            if k in gene2path:\n",
    "                gene2path[k] += [j]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "gene2path is a dictionary mapping the genes with a linear relationship to age with the pathways they're involved in. Let's use this to find the genes that are most associated with the enriched pathways.\n",
    "\n",
    "We'll start by eliminating genes that have 1 or less significant pathways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possibletargets = { i : j for i,j in gene2path.items() if len(j) > 1 }\n",
    "possibletargets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Those are our 2 targets. Let's look at the data over time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Genes, Cells) = (0, 221)\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py:3361\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3359'>3360</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3360'>3361</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3361'>3362</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/_libs/index.pyx:76\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/_libs/index.pyx:108\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5198\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5206\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Gene'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb Cell 30'\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=4'>5</a>\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(folder\u001b[39m+\u001b[39mf,delimiter\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m'\u001b[39m)[ [i \u001b[39min\u001b[39;00m possibletargets \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m df[\u001b[39m'\u001b[39m\u001b[39mGene\u001b[39m\u001b[39m'\u001b[39m]] ]\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=5'>6</a>\u001b[0m counts \u001b[39m=\u001b[39m df[df\u001b[39m.\u001b[39mcolumns[\u001b[39m1\u001b[39m:]]\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=6'>7</a>\u001b[0m genes \u001b[39m=\u001b[39m df[\u001b[39m'\u001b[39;49m\u001b[39mGene\u001b[39;49m\u001b[39m'\u001b[39;49m]\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=7'>8</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39m(Genes, Cells) =\u001b[39m\u001b[39m'\u001b[39m,counts\u001b[39m.\u001b[39mshape)\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=8'>9</a>\u001b[0m genemean \u001b[39m=\u001b[39m counts\u001b[39m.\u001b[39mmean(axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py:3458\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3455'>3456</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mnlevels \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3456'>3457</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3457'>3458</a>\u001b[0m indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(key)\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3458'>3459</a>\u001b[0m \u001b[39mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3459'>3460</a>\u001b[0m     indexer \u001b[39m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py:3363\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3360'>3361</a>\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3361'>3362</a>\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m-> <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3362'>3363</a>\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3364'>3365</a>\u001b[0m \u001b[39mif\u001b[39;00m is_scalar(key) \u001b[39mand\u001b[39;00m isna(key) \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhasnans:\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3365'>3366</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Gene'"
     ]
    }
   ],
   "source": [
    "folder = 'DenoisedData/'\n",
    "stds = []\n",
    "means = []\n",
    "for i,f in enumerate(sorted(listdir(folder))):\n",
    "    df = pd.read_csv(folder+f,delimiter='\\t')[ [i in possibletargets for i in df['Gene']] ]\n",
    "    counts = df[df.columns[1:]]\n",
    "    genes = df['Gene']\n",
    "    print('(Genes, Cells) =',counts.shape)\n",
    "    genemean = counts.mean(axis=1)\n",
    "    genestd = counts.std(axis=1)\n",
    "\n",
    "    \n",
    "    stds += [genestd]\n",
    "    means += [genemean]\n",
    "\n",
    "stds = np.array(list(stds))\n",
    "means = np.array(list(means))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.363897847707054 44.275400568066885\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXIAAAD5CAYAAAA6JL6mAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAP+ElEQVR4nO3df2yd1X3H8c+HOEChWWHEMEbIzDo0rUAbJI+2oqgdo1s2UGGTKujataompUhbSauOQpmWUdROItNQxdRpRC0FqRk/WmBDJdMKaQmLxAAHQsEERGgJhCTYARLMWhLH/u6Pew1O7OAb3+P73O+975dk5d7Hl3PPCc7H33ue85zHESEAQF6HVd0BAEBzCHIASI4gB4DkCHIASI4gB4DkCHIASK6nVEO250kakPRSRFzwTq9duHBh9PX1lXprAOgKGzZs2BkRvQceLxbkkpZL2iTp12Z6YV9fnwYGBgq+NQB0PttbpjteZGrF9iJJ50v6Ton2AACNKzVH/i1JX5U0Xqg9AECDmg5y2xdIGoqIDTO8bpntAdsDw8PDzb4tAKCuREV+tqRP2H5e0q2SzrX9/QNfFBGrIqI/Ivp7e6fM1QMAZqnpII+Ir0XEoojok3SJpJ9ExGea7hkAoCGsIweA5EouP1RE3C/p/pJtAgDeGRU5ACRHkAPd7Hvn176QGkEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACqx71Dm0KQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAtMocXcFKkANAck0Hue0jbT9s+3Hbg7a/XqJjAIDG9BRoY4+kcyPiDdvzJa23/V8R8b8F2gYAzKDpII+IkPRG/en8+lc02y4AoDFF5shtz7O9UdKQpHsj4qES7QJAJxncvluD23cXb7dIkEfEWEQskbRI0lm2Tz/wNbaX2R6wPTA8PFzibQEAKrxqJSJ2Sbpf0tJpvrcqIvojor+3t7fk2wJAVyuxaqXX9jH1x++SdJ6kp5ttF0D3mKsph25RYtXKiZJutj1PtV8Mt0fEjwq0CwBoQIlVKz+TdGaBvgAAZoErOwEgOYIcAJIjyAEguRInOwEkNbFS5LSK+4HmUJEDQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5DP4OIbHtTFNzxYdTcA4KAIcgBIjiAHgOQIcgBIjiAHgOQIcgBIjiAHgOQI8gnfO7/2BQDJEOQAkBxBDgDJEeQAkBxBjkqw9QFQDkEOAMkR5ACQHEEOAMkR5ACQHEEOAMk1HeS2T7b9U9ubbA/aXl6iYwCAxvQUaGOfpK9ExKO2F0jaYPveiHiqQNsAgBk0XZFHxPaIeLT+eETSJkknNdsuAKAxRefIbfdJOlPSQyXbBQAcXLEgt/1uSXdI+lJEvD7N95fZHrA9MDw8XOpt0WG44hM4dEWC3PZ81UJ8dUTcOd1rImJVRPRHRH9vb2+JtwUAqMyqFUv6rqRNEXFd810CAByKEqtWzpb0l5KesL2xfuyqiFhToG10mRWvXF5/tL7SfgCZNB3kEbFekgv0BQAwC6mu7OREGABMlSrIAQBTEeQAkBxBDgDJEeQAkBxBDqByPTGqvtHnpJGXq+5KSgQ5gMotHBvSUfFLad21VXclJYIc6GJtUQmP7NCx46/VLkbZuJqqfBYIcqCLtUUlvG6lpKg9jnGq8lkgyIFu1Q6V8MgOaePqt4NobG9HV+UjY/M1/uYb2rnjhaLtEuRAt2qHSnjdytp7T9bBVfme0TGd5ue1+QcrirZLkAPdqF0q4a0P1957srG9teMdZue2Lfqgn9JhDi3ZeU/RqpwgB7pRu1TCl66Xrt6twcPP0ODhZ0hX7659Xdp5u18+d8cKuf4JyBovWpUT5EA36qJKuB3s3LZFH9h5j47wPknSEd5XtCovsR85gGzqFe/gP35EknTaVZ1XAbeT5+5YoSUT5yPqJqryhV+8qen2qcgBYI4t3PWzt6rxCUd4n3p3PV6kfSpyAJV7dfxoXfuri3TjyJs6fsGRVXenuPf+/WOSpn4Cem+h9qnIAVTulj1na3Bska5fu7nqrqREkAOo1NDrb+q+0fcrdJh+OPCihkberLpL6RDkACp1/dpnNV6/7e9YBFX5LBDkACoz9Pqb+sGGrdpXP103OhZU5bNAkAOozPVrn9V47L8sj6r80BHkACrz6Au7NDq2f5CPjoUe3fJaRT3KieWHACqzZvk5krgwqVkEOSqx4pXL64/4hws0i6kVAEiOIAe62KvjR+uK//s0q0SSI8iBLsYVlZ2BIAe6FFdUdg6CHOhSXFHZOQhyoAtxRWVnIcjrBrfv1uD23VOO7903rqe2vc4PODoKV1R2FoJ8Bi/t+pVG9uzjBxwdhSsqOwsXBNX1xKgW7XuhdhfxBSdIqn38HH5jjyTphwMv6rI//J2O3PQe3YcrKjtLkYrc9o22h2w/WaK9KiwcG9JR8cv97iJ+/dpnNXGbPT52AmhXpaZWbpK0tFBbrTeyQ8eOv1Y7f79xtTTy8lsngyY+fHIyqDV6YlR9o8/VPhkBaEiRII+IByS9WqKtSqxbqbdK7xiX1l3LyaCKTPfJCMA7a9nJTtvLbA/YHhgeHp5VGyteuXzSZkuFjOyQNq5++y9ibK+0cbV+8fwvOBnUaiM7dEz9k1HUPxkBmFnLTnZGxCpJqySpv78/Znh566xbWavCJ4txrT71funL1+niGx6UJN32hQ+3vm/dZt1Kjcdhmucxje0bU8+6a6ULrqu6V0DbY/nh1odrVfhkY3trxzFnpmzWNLJD8dhqzfeYJKlHo1TlQINYfnhpbdnVwZZhsW/23Ji8WdM3LjpdWrdSY2Nj+/1AUpUDjSm1/PAWSQ9K+l3bW23/VYl20Zmm26xpdMtD6tHofq/r0ahGtzxUUS+BPIpU5BHxqRLtoDtMu1nTb/6bbtv24n4nmOfPsy4+abG+UVVHgSSYWkFLHWyzpsXHHcUqIWCWCHK01MHW5591ynH68Zc/yiXjwCywagUtxWZNQHlU5GgpNmsCyqMiB4DkCHIASI4gB4DkCHIASI4gB4DkCHJUghtIAOUQ5KgEN5AAykkV5FO2PkVO09xaD8DspQryyVufIrFpbq0HYPbSBPl0W58ioYPcWo+qHJi9NEE+7danyOcgt9ajKgdmL0WQH2zrU6ryhLi1HlBciiCf2Pq0V6/ptsOvUa92tawqZ5lcYZeul67ercHDz9Dg4WdIV++ufV3K5lnAbKUI8omtTy/ruUu/72f0xZ47W7b1KcvkALS7FNvYrll+jjSyQ+P//GkdptBnj1yvzy7/V2nBCXP7xgcuk/voFXP/ngBwiFJU5JLmfMnatGvUWSYHIIEcQd6CJWtT1qizTA5AEjmCfI6XrE27Rp1lcgCSyBHkc7xkbdo16iyTA5BEipOdE0vT5uI+jwdbo37ZFffp+AVHcm9JAG0vR0U+hybWqE/GlaMAMun6IJ9Yoz5Zq9aoA0AJOaZW5tCa5edImptpGwBoha6vyAEgu66vyAGgVU478T1z0i4VOQAkR5ADQHIEOQAkR5ADQHJFgtz2UtvP2N5s+8oSbQIAGtP0qhXb8yR9W9LHJW2V9IjtuyPiqWbbBoCO8vl75qTZEhX5WZI2R8TPI2KvpFslXVigXQBAA0qsIz9J0ouTnm+V9MEDX2R7maRlkrR48eICb4tOdM1x/yRJuq3ifgCZlAhyT3MsphyIWCVplST19/dP+T6A7sUv8OaUmFrZKunkSc8XSdpWoF0AQANKVOSPSDrV9imSXpJ0iaS/KNAugC5x2xc+XHUXUms6yCNin+2/kfTfkuZJujEiBpvuGQCgIUU2zYqINZLWlGir3TB3B6DdsfshKnGwX5B8xAYOHZfoA0ByqSpypjkAYCoqcgBIjiAHgOQIcgBILtUc+Vxi/h1AVlTkAJAcQQ4AyTG1MgMuUAHQ7qjIASA5ghwAkmNqBehip534nqq7gAKoyAEgOSryOk5qthZ/30A5VOQAkBxBDgDJMbUCdLPP31N1D1AAFTkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByqTbNYg9rAJiKihwAkiPIASC5poLc9idtD9oet91fqlMAgMY1W5E/KenPJT1QoC8AgFlo6mRnRGySJNtlegMAOGQtmyO3vcz2gO2B4eHhVr0tAHS8GSty2/dJ+o1pvvV3EfGfjb5RRKyStEqS+vv7o+EeAgDe0YxBHhHntaIjAIDZYfkhACTniNnPctj+M0n/IqlX0i5JGyPijxv474YlbZnhZQsl7Zx15/LohnF2wxil7hhnN4xRat9x/lZE9B54sKkgn0u2ByKi49emd8M4u2GMUneMsxvGKOUbJ1MrAJAcQQ4AybVzkK+qugMt0g3j7IYxSt0xzm4Yo5RsnG07Rw4AaEw7V+QAgAa0ZZDbXmr7GdubbV9ZdX9KsX2j7SHbT0469uu277X9bP3PY6vsY7Nsn2z7p7Y31XfGXF4/3jHjtH2k7YdtP14f49frxztmjBNsz7P9mO0f1Z934hift/2E7Y22B+rHUo2z7YLc9jxJ35b0J5LeJ+lTtt9Xba+KuUnS0gOOXSlpbUScKmlt/Xlm+yR9JSJ+T9KHJP11/f9fJ41zj6RzI+IDkpZIWmr7Q+qsMU5YLmnTpOedOEZJ+oOIWDJpyWGqcbZdkEs6S9LmiPh5ROyVdKukCyvuUxER8YCkVw84fKGkm+uPb5Z0USv7VFpEbI+IR+uPR1QLgZPUQeOMmjfqT+fXv0IdNEZJsr1I0vmSvjPpcEeN8R2kGmc7BvlJkl6c9Hxr/VinOiEitku1EJR0fMX9KcZ2n6QzJT2kDhtnfcpho6QhSfdGRMeNUdK3JH1V0vikY502Rqn2S/jHtjfYXlY/lmqc7Xjz5ek2N2dpTTK23y3pDklfiojXO23P+ogYk7TE9jGS7rJ9esVdKsr2BZKGImKD7Y9V3J25dnZEbLN9vKR7bT9ddYcOVTtW5FslnTzp+SJJ2yrqSyu8bPtESar/OVRxf5pme75qIb46Iu6sH+64cUpSROySdL9q5z46aYxnS/qE7edVm9481/b31VljlCRFxLb6n0OS7lJtejfVONsxyB+RdKrtU2wfLukSSXdX3Ke5dLekz9Uff05Sw3u8tyPXSu/vStoUEddN+lbHjNN2b70Sl+13STpP0tPqoDFGxNciYlFE9Kn2b/AnEfEZddAYJcn20bYXTDyW9Eeq3cIy1Tjb8oIg23+q2vzcPEk3RsQ3q+1RGbZvkfQx1XZWe1nSP0j6D0m3S1os6QVJn4yIA0+IpmH7I5L+R9ITentu9SrV5sk7Ypy236/aCbB5qhVDt0fENbaPU4eMcbL61MrfRsQFnTZG27+tWhUu1aaa/z0ivpltnG0Z5ACAxrXj1AoA4BAQ5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQ3P8DvoFTovhHOmoAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "t = [int(i[:2]) for i in sorted(listdir(folder))]\n",
    "plt.errorbar(t, means[:,0]/np.median(means[:,0]), stds[:,0]/np.median(means[:,0]), linestyle='None', marker='^')\n",
    "plt.errorbar(t, means[:,1]/np.median(means[:,1]), stds[:,1]/np.median(means[:,1]), linestyle='None', marker='^')\n",
    "print(np.std(means[:,1]),max(stds[:,1]))\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "569d0bbc3854db4a73f5be0bb5576848176d13611564839ca8e34057b1299597"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
