{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "from os import listdir\n",
    "import shutil\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from matplotlib import colors\n",
    "from sklearn.manifold import TSNE\n",
    "import subprocess\n",
    "import shlex\n",
    "from sklearn.metrics import r2_score\n",
    "import random\n",
    "\n",
    "# todo:\n",
    "# https://www.embopress.org/doi/full/10.15252/msb.20188746\n",
    "# https://scanpy.readthedocs.io/en/stable/tutorials.html\n",
    "# umap / tsne on the whole dataset to see if ages are naturally seperable\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aging is the time dependant weaking of the body. For example, aging in the pancreous can predispose people to diabetes (https://www.imrpress.com/journal/FBE/10/1/10.2741/e813). I will attempt to find theraputic targets that could be used to target aging in the pancreatus using scRNA-Seq data from  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81547.\n",
    "\n",
    "A good target is a gene or pathway that:\n",
    "\n",
    "(1) Has reletively constant expression accross cells from the same individual. Targeted cell delivery is a hard process, especially between cell types in the same organ, so the genes or pathways need to be consistant accross the single cells.\n",
    "\n",
    "(2) Is highly corralative with age. This is less important because if there is a gene whose expression is dependant on age in some nonlinear maner, it could still be a good target if it is modulated carefully. This is a simplifying assumption.\n",
    "\n",
    "(3) Has a large magnatude of change with age. Something that only changes subtley with age will be harder to modulate. This is more for gene selection rather than pathway selection."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Preprocessing\n",
    "\n",
    "Creates the aggragated data for each sample in the ClupedData folder\n",
    "\n",
    "This just puts the data in a nicer form for downstream processing: files named with the donor age and sex with the counts of each RNA in each row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the description of the data:\n",
    "\n",
    "\tAs organisms age, cells accumulate genetic and epigenetic changes that eventually lead to impaired organ function or catastrophic failure such as cancer. Here we describe a single-cell transcriptome analysis of 2544 human pancreas cells from donors, spanning six decades of life. We find that islet cells from older donors have increased levels of disorder as measured both by noise in the transcriptome and by the number of cells which display inappropriate hormone expression, revealing a transcriptional instability associated with aging. By analyzing the spectrum of somatic mutations in single cells from previously-healthy donors, we find a specific age-dependent mutational signature characterized by C to A and C to G transversions, indicators of oxidative stress, which is absent in single cells from human brain tissue or in a tumor cell line. Cells carrying a high load of such mutations also express higher levels of stress and senescence markers, including FOS, JUN, and the cytoplasmic superoxide dismutase SOD1, markers previously linked to pancreatic diseases with substantial age-dependent risk, such as type 2 diabetes mellitus and adenocarcinoma. Thus, our single-cell approach unveils gene expression changes and somatic mutations acquired in aging human tissue, and identifies molecular pathways induced by these genetic changes that could influence human disease. Also, our results demonstrate the feasibility of using single-cell RNA-seq data from primary cells to derive meaningful insights into the genetic processes that operate on aging human tissue and to determine which molecular mechanisms are coordinated with these processes.\n",
    "\n",
    "This dataset contains scRNA-seq data on the pancreas of humans of various ages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1/2-SBSRNA4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2LD1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23460</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>407558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23461</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>3754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23462</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23463</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23464</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>85563</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23465 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0       1\n",
       "0               1/2-SBSRNA4       0\n",
       "1                      A1BG       0\n",
       "2                  A1BG-AS1       0\n",
       "3                      A1CF       0\n",
       "4                     A2LD1       0\n",
       "...                     ...     ...\n",
       "23460            no_feature  407558\n",
       "23461             ambiguous    3754\n",
       "23462         too_low_aQual       0\n",
       "23463           not_aligned       0\n",
       "23464  alignment_not_unique   85563\n",
       "\n",
       "[23465 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = 'GSE81547'\n",
    "datasets = sorted([path+'/'+i for i in listdir(path)])\n",
    "\n",
    "pd.read_csv(datasets[0],delimiter='\\t',header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The files themselves don't contain any information on the samples. Clumn 0 contains the gene and column 1 contains the Transcript counts. The last 5 rows contains data on the RNA-Seq preprocessing. First lets use the metadata file to get the ages and sexes of each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "id2file = lambda id: [i for i in datasets if i[i.find('/')+1:i.find('_')] == id][0]\n",
    "with open('GSE81547_series_matrix.txt') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('!Series_sample_id'):\n",
    "            identifier = line[len('!Series_sample_id\\t')+1:-3].split(' ')\n",
    "        if line.startswith('!Sample_title'):\n",
    "            donor = [(int(i[1:i.find('yr')]),i[i.find('_')+1]) for i in line[len('!Sample_title\\t'):].split('\\t')]\n",
    "            donor = [(i, [id2file(identifier[j]) for j,k in enumerate(donor) if k == i]) for i in set(donor)]\n",
    "donor.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for id,files in donor:\n",
    "    dfs = [pd.read_csv(files[0],delimiter='\\t',header=None)] + [pd.read_csv(f,delimiter='\\t',header=None)[1] for f in files[1:]]\n",
    "    totdata = pd.concat(dfs,axis=1).iloc[:-5, :]\n",
    "    totdata.columns = ['Gene'] + list(range(len(totdata.columns)-1))\n",
    "    totdata.to_csv(f'ClumpedData/{\"0\"*(id[0]<10)}{id[0]}_{id[1]}.csv',index=False,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have a function that takes the file name and returns the metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Cleaning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will be using enhance from https://www.biorxiv.org/content/10.1101/655365v2.full\n",
    "\n",
    "The algorithm is incredibly simple. It basically assumes the data has a large nullspace and the data is noisy and uses PCA to choose which dimensions to keep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = 'ClumpedData/'\n",
    "for file in listdir(folder):\n",
    "    subprocess.call(shlex.split(f'python3 enhance/enhance.py -f {folder+file} -o DenoisedData/{file} --seed 0'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the data is denoised, I can start removing genes that won't help me find a theraputic target for aging.\n",
    "\n",
    "The first thing to do would be removing genes whose standard deviation is larger within groups that throughout groups. The only reason this is valid is because I already cleaned the data to (hopefully) remove the technical noise. I will also log transform the data, as is standard with scRNA seq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Genes, Cells) = (23460, 221)\n",
      "(Genes, Cells) = (23460, 331)\n",
      "(Genes, Cells) = (23460, 178)\n",
      "(Genes, Cells) = (23460, 505)\n",
      "(Genes, Cells) = (23460, 286)\n",
      "(Genes, Cells) = (23460, 474)\n",
      "(Genes, Cells) = (23460, 277)\n",
      "(Genes, Cells) = (23460, 272)\n",
      "Failed genes due to high in group variance: 16102\n",
      "Possible genes: 7358\n"
     ]
    }
   ],
   "source": [
    "folder = 'DenoisedData/'\n",
    "stds = []\n",
    "means = []\n",
    "for i,f in enumerate(sorted(listdir(folder))):\n",
    "    df = pd.read_csv(folder+f,delimiter='\\t')\n",
    "    counts = df[df.columns[1:]]\n",
    "    genes = df['Gene']\n",
    "    print('(Genes, Cells) =',counts.shape)\n",
    "    genemean = counts.mean(axis=1)\n",
    "    genestd = counts.std(axis=1)\n",
    "\n",
    "    \n",
    "    stds += [genestd]\n",
    "    means += [genemean]\n",
    "\n",
    "stds = np.array(list(stds))\n",
    "means = np.array(list(means))\n",
    "outgroupstd = means.std(axis=0)\n",
    "ingroupstd = np.median(stds,axis=0)\n",
    "print('Failed genes due to high in group variance:',sum(ingroupstd >= outgroupstd))\n",
    "print('Possible genes:',sum(ingroupstd < outgroupstd))\n",
    "potentialgenes = genes[ingroupstd < outgroupstd]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genes Monotonically Corrolated with Age"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's search for any genes that highly corrolate with age. They would be potential targets\n",
    "\n",
    "I use the Spearmans r and p values to find genes that are associated with aging. I am using Spearmanr because I only care that the relashionship is monotonic, not that it is linear."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corrolated genes that also work with the spearmons corrolation criteria: 37\n",
      "AGT\tANKRD26P1\tARVCF\tBAI2\tCCDC65\tCDCA7\tCHRDL2\tCRH\tDGKQ\tDLGAP3\tDLL3\tFIZ1\tGAB3\tHIST1H1D\tJRK\tKIAA1644\tLCN10\tLEF1\tLOC100506134\tLOC389332\tMYBPHL\tORM1\tPCK2\tPID1\tPRKG1\tRBM24\tSAA1\tSHISA7\tSIX3\tSMOC1\tSUGT1P3\tTAF10\tTFF1\tTMEM176B\tTOR2A\tTSIX\tVGF\t"
     ]
    }
   ],
   "source": [
    "# print(means.shape)\n",
    "t = [int(i[:2]) for i in sorted(listdir(folder))]\n",
    "goodinds = []\n",
    "for i,y in enumerate(means[:,potentialgenes.index].T):\n",
    "    if np.max(y) > 1:\n",
    "        r,p = sp.stats.spearmanr(t,y)\n",
    "        if abs(r) > 0.8 and p < 0.05:\n",
    "            goodinds += [potentialgenes.index[i]]\n",
    "\n",
    "\n",
    "print('Corrolated genes that also work with the spearmons corrolation criteria:',len(goodinds))\n",
    "with open('TargetGenesV1.txt','w+') as f:\n",
    "    for i in genes[goodinds]:\n",
    "        print(i,end='\\t')\n",
    "        f.write(i+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reactome Pathway Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aging has many causes, so targeting any specific gene will likely not result in a succesful drug. I will see which pathways are differentially regulated due to age so they could be used as a drug target. The idea is if, for example, immune system pathways are enriched, immunosuprescence could be a potential drug target.\n",
    "\n",
    "\n",
    "I will be using Reactome for my pathway analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Section 4 shows the most upregulated pathways. There's plenty of pathways for a biologist to interperate. The most important thing to look at is the False Discovery Rate (FDR). It can be thought of as the probability that the pathway is differentially expressed through chance alone, as opposed to through biologically relevent meanings.\n",
    "\n",
    "For a visual interprotation, see the image below.\n",
    "\n",
    "The more yellow a pathway is, the more confident Reactome is that relates to aging. For example, genes associated with Apoptosis Induces DNA Fragmentation are significantly associated with aging in the pancreous.\n",
    "\n",
    "![alt text](ReacfoamV1.jpg \"PathwaysV0\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlap\n",
    "\n",
    "This section does not depend on variables from previous sections"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's see which of these pathways could be targeted using the gene list we found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway identifier</th>\n",
       "      <th>Pathway name</th>\n",
       "      <th>#Entities found</th>\n",
       "      <th>#Entities total</th>\n",
       "      <th>#Interactors found</th>\n",
       "      <th>#Interactors total</th>\n",
       "      <th>Entities ratio</th>\n",
       "      <th>Entities pValue</th>\n",
       "      <th>Entities FDR</th>\n",
       "      <th>#Reactions found</th>\n",
       "      <th>#Reactions total</th>\n",
       "      <th>Reactions ratio</th>\n",
       "      <th>Species identifier</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Submitted entities found</th>\n",
       "      <th>Mapped entities</th>\n",
       "      <th>Submitted entities hit interactor</th>\n",
       "      <th>Interacts with</th>\n",
       "      <th>Found reaction identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-140342</td>\n",
       "      <td>Apoptosis induced DNA fragmentation</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000863</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>0.362959</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000879</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-HSA-211239;R-HSA-211247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R-HSA-9022702</td>\n",
       "      <td>MECP2 regulates transcription of neuronal ligands</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0.000863</td>\n",
       "      <td>0.002321</td>\n",
       "      <td>0.450203</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000586</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CRH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-HSA-9017983;R-HSA-9017988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R-HSA-444473</td>\n",
       "      <td>Formyl peptide receptors bind formyl peptides ...</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000730</td>\n",
       "      <td>0.043308</td>\n",
       "      <td>0.550449</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SAA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-HSA-391913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R-HSA-2559584</td>\n",
       "      <td>Formation of Senescence-Associated Heterochrom...</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>116</td>\n",
       "      <td>0.001128</td>\n",
       "      <td>0.044139</td>\n",
       "      <td>0.550449</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-HSA-4647594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R-HSA-1296061</td>\n",
       "      <td>HCN channels</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.045797</td>\n",
       "      <td>0.550449</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRKG1</td>\n",
       "      <td>17489</td>\n",
       "      <td>R-HSA-1297444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>R-HSA-73857</td>\n",
       "      <td>RNA Polymerase II Transcription</td>\n",
       "      <td>4</td>\n",
       "      <td>1695</td>\n",
       "      <td>14</td>\n",
       "      <td>4788</td>\n",
       "      <td>0.112468</td>\n",
       "      <td>0.999971</td>\n",
       "      <td>0.999971</td>\n",
       "      <td>45</td>\n",
       "      <td>891</td>\n",
       "      <td>0.065294</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>TAF10;LEF1;CRH</td>\n",
       "      <td>Q12962;ENSG00000147571;Q9UJU2;P06850</td>\n",
       "      <td>TAF10;TFF1;PCK2;JRK;ORM1;LEF1;AGT;PID1;SIX3;SM...</td>\n",
       "      <td>Q92831;P03372;P10809;O14744;P05121;O75604;Q925...</td>\n",
       "      <td>R-HSA-8853956;R-HSA-109637;R-HSA-109636;R-HSA-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>R-HSA-1430728</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>4</td>\n",
       "      <td>3643</td>\n",
       "      <td>5</td>\n",
       "      <td>2671</td>\n",
       "      <td>0.241723</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>13</td>\n",
       "      <td>2251</td>\n",
       "      <td>0.164957</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ORM1;AGT;PCK2</td>\n",
       "      <td>Q16822;P01019;ENSG00000135744;Q8N138</td>\n",
       "      <td>KIAA1644;TMEM176B;JRK;PRKG1;AGT</td>\n",
       "      <td>P13196;P17252;Q8IY26;Q8N0U8;17489</td>\n",
       "      <td>R-HSA-8952137;R-HSA-381668;R-HSA-381607;R-HSA-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>R-HSA-74160</td>\n",
       "      <td>Gene expression (Transcription)</td>\n",
       "      <td>4</td>\n",
       "      <td>1858</td>\n",
       "      <td>14</td>\n",
       "      <td>4996</td>\n",
       "      <td>0.123283</td>\n",
       "      <td>0.999987</td>\n",
       "      <td>0.999987</td>\n",
       "      <td>47</td>\n",
       "      <td>1006</td>\n",
       "      <td>0.073721</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>TAF10;LEF1;CRH</td>\n",
       "      <td>Q12962;ENSG00000147571;Q9UJU2;P06850</td>\n",
       "      <td>TAF10;TFF1;PCK2;JRK;ORM1;LEF1;AGT;PID1;SIX3;SM...</td>\n",
       "      <td>Q92831;P03372;P10809;O14744;P05121;O75604;Q925...</td>\n",
       "      <td>R-HSA-73729;R-HSA-8853956;R-HSA-109637;R-HSA-1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>R-HSA-1266738</td>\n",
       "      <td>Developmental Biology</td>\n",
       "      <td>1</td>\n",
       "      <td>1263</td>\n",
       "      <td>5</td>\n",
       "      <td>3113</td>\n",
       "      <td>0.083803</td>\n",
       "      <td>0.999990</td>\n",
       "      <td>0.999990</td>\n",
       "      <td>10</td>\n",
       "      <td>556</td>\n",
       "      <td>0.040745</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>LEF1</td>\n",
       "      <td>Q9UJU2</td>\n",
       "      <td>KIAA1644;TMEM176B;TFF1;VGF;CHRDL2</td>\n",
       "      <td>P05412;P17252;Q01453;P49639;O75093</td>\n",
       "      <td>R-HSA-9622386;R-HSA-9616243;R-HSA-9010396;R-HS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>R-HSA-1643685</td>\n",
       "      <td>Disease</td>\n",
       "      <td>2</td>\n",
       "      <td>2435</td>\n",
       "      <td>8</td>\n",
       "      <td>4743</td>\n",
       "      <td>0.161569</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>33</td>\n",
       "      <td>1628</td>\n",
       "      <td>0.119302</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>TAF10;CRH</td>\n",
       "      <td>Q12962;P06850</td>\n",
       "      <td>TFF1;VGF;PRKG1;PID1;AGT;DLGAP3;AGT;MYBPHL</td>\n",
       "      <td>P05412;Q08379;Q9BYF1;Q15078;P27797;17489</td>\n",
       "      <td>R-HSA-169921;R-HSA-392129;R-HSA-9660824;R-HSA-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>335 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Pathway identifier                                       Pathway name  \\\n",
       "0         R-HSA-140342                Apoptosis induced DNA fragmentation   \n",
       "1        R-HSA-9022702  MECP2 regulates transcription of neuronal ligands   \n",
       "2         R-HSA-444473  Formyl peptide receptors bind formyl peptides ...   \n",
       "3        R-HSA-2559584  Formation of Senescence-Associated Heterochrom...   \n",
       "4        R-HSA-1296061                                       HCN channels   \n",
       "..                 ...                                                ...   \n",
       "330        R-HSA-73857                    RNA Polymerase II Transcription   \n",
       "331      R-HSA-1430728                                         Metabolism   \n",
       "332        R-HSA-74160                    Gene expression (Transcription)   \n",
       "333      R-HSA-1266738                              Developmental Biology   \n",
       "334      R-HSA-1643685                                            Disease   \n",
       "\n",
       "     #Entities found  #Entities total  #Interactors found  #Interactors total  \\\n",
       "0                  2               13                   0                   6   \n",
       "1                  2               13                   0                  14   \n",
       "2                  1               11                   0                   7   \n",
       "3                  2               17                   0                 116   \n",
       "4                  0                6                   1                  13   \n",
       "..               ...              ...                 ...                 ...   \n",
       "330                4             1695                  14                4788   \n",
       "331                4             3643                   5                2671   \n",
       "332                4             1858                  14                4996   \n",
       "333                1             1263                   5                3113   \n",
       "334                2             2435                   8                4743   \n",
       "\n",
       "     Entities ratio  Entities pValue  Entities FDR  #Reactions found  \\\n",
       "0          0.000863         0.000935      0.362959                 2   \n",
       "1          0.000863         0.002321      0.450203                 2   \n",
       "2          0.000730         0.043308      0.550449                 1   \n",
       "3          0.001128         0.044139      0.550449                 1   \n",
       "4          0.000398         0.045797      0.550449                 1   \n",
       "..              ...              ...           ...               ...   \n",
       "330        0.112468         0.999971      0.999971                45   \n",
       "331        0.241723         0.999977      0.999977                13   \n",
       "332        0.123283         0.999987      0.999987                47   \n",
       "333        0.083803         0.999990      0.999990                10   \n",
       "334        0.161569         1.000000      1.000000                33   \n",
       "\n",
       "     #Reactions total  Reactions ratio  Species identifier  Species name  \\\n",
       "0                  12         0.000879                9606  Homo sapiens   \n",
       "1                   8         0.000586                9606  Homo sapiens   \n",
       "2                   3         0.000220                9606  Homo sapiens   \n",
       "3                   2         0.000147                9606  Homo sapiens   \n",
       "4                   2         0.000147                9606  Homo sapiens   \n",
       "..                ...              ...                 ...           ...   \n",
       "330               891         0.065294                9606  Homo sapiens   \n",
       "331              2251         0.164957                9606  Homo sapiens   \n",
       "332              1006         0.073721                9606  Homo sapiens   \n",
       "333               556         0.040745                9606  Homo sapiens   \n",
       "334              1628         0.119302                9606  Homo sapiens   \n",
       "\n",
       "    Submitted entities found                       Mapped entities  \\\n",
       "0                   HIST1H1D                                   NaN   \n",
       "1                        CRH                                   NaN   \n",
       "2                       SAA1                                   NaN   \n",
       "3                   HIST1H1D                                   NaN   \n",
       "4                        NaN                                   NaN   \n",
       "..                       ...                                   ...   \n",
       "330           TAF10;LEF1;CRH  Q12962;ENSG00000147571;Q9UJU2;P06850   \n",
       "331            ORM1;AGT;PCK2  Q16822;P01019;ENSG00000135744;Q8N138   \n",
       "332           TAF10;LEF1;CRH  Q12962;ENSG00000147571;Q9UJU2;P06850   \n",
       "333                     LEF1                                Q9UJU2   \n",
       "334                TAF10;CRH                         Q12962;P06850   \n",
       "\n",
       "                     Submitted entities hit interactor  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                PRKG1   \n",
       "..                                                 ...   \n",
       "330  TAF10;TFF1;PCK2;JRK;ORM1;LEF1;AGT;PID1;SIX3;SM...   \n",
       "331                    KIAA1644;TMEM176B;JRK;PRKG1;AGT   \n",
       "332  TAF10;TFF1;PCK2;JRK;ORM1;LEF1;AGT;PID1;SIX3;SM...   \n",
       "333                  KIAA1644;TMEM176B;TFF1;VGF;CHRDL2   \n",
       "334          TFF1;VGF;PRKG1;PID1;AGT;DLGAP3;AGT;MYBPHL   \n",
       "\n",
       "                                        Interacts with  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                17489   \n",
       "..                                                 ...   \n",
       "330  Q92831;P03372;P10809;O14744;P05121;O75604;Q925...   \n",
       "331                  P13196;P17252;Q8IY26;Q8N0U8;17489   \n",
       "332  Q92831;P03372;P10809;O14744;P05121;O75604;Q925...   \n",
       "333                 P05412;P17252;Q01453;P49639;O75093   \n",
       "334           P05412;Q08379;Q9BYF1;Q15078;P27797;17489   \n",
       "\n",
       "                            Found reaction identifiers  \n",
       "0                            R-HSA-211239;R-HSA-211247  \n",
       "1                          R-HSA-9017983;R-HSA-9017988  \n",
       "2                                         R-HSA-391913  \n",
       "3                                        R-HSA-4647594  \n",
       "4                                        R-HSA-1297444  \n",
       "..                                                 ...  \n",
       "330  R-HSA-8853956;R-HSA-109637;R-HSA-109636;R-HSA-...  \n",
       "331  R-HSA-8952137;R-HSA-381668;R-HSA-381607;R-HSA-...  \n",
       "332  R-HSA-73729;R-HSA-8853956;R-HSA-109637;R-HSA-1...  \n",
       "333  R-HSA-9622386;R-HSA-9616243;R-HSA-9010396;R-HS...  \n",
       "334  R-HSA-169921;R-HSA-392129;R-HSA-9660824;R-HSA-...  \n",
       "\n",
       "[335 rows x 19 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('ReactomeDataV1.csv')\n",
    "genes = set(pd.read_csv('TargetGenesV1.txt',header=None)[0])\n",
    "\n",
    "gene2path = { i : [] for i in genes}\n",
    "for i , j , fdr in zip(data['Submitted entities found'],data['Pathway name'],data['Entities FDR']):\n",
    "    if fdr < 0.05:\n",
    "        for k in i.strip().split(';'):\n",
    "            if k in gene2path:\n",
    "                gene2path[k] += [j]\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "gene2path is a dictionary mapping the genes with a linear relationship to age with the pathways they're involved in. Let's use this to find the genes that are most associated with the enriched pathways.\n",
    "\n",
    "We'll start by eliminating genes that have 1 or less significant pathways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ANKRD26P1': [],\n",
       " 'BAI2': [],\n",
       " 'DGKQ': [],\n",
       " 'HIST1H1D': [],\n",
       " 'PID1': [],\n",
       " 'LCN10': [],\n",
       " 'CHRDL2': [],\n",
       " 'CCDC65': [],\n",
       " 'ARVCF': [],\n",
       " 'KIAA1644': [],\n",
       " 'TFF1': [],\n",
       " 'LOC389332': [],\n",
       " 'SHISA7': [],\n",
       " 'LOC100506134': [],\n",
       " 'JRK': [],\n",
       " 'SUGT1P3': [],\n",
       " 'LEF1': [],\n",
       " 'AGT': [],\n",
       " 'MYBPHL': [],\n",
       " 'TMEM176B': [],\n",
       " 'PCK2': [],\n",
       " 'FIZ1': [],\n",
       " 'ORM1': [],\n",
       " 'RBM24': [],\n",
       " 'CDCA7': [],\n",
       " 'TAF10': [],\n",
       " 'VGF': [],\n",
       " 'TSIX': [],\n",
       " 'SMOC1': [],\n",
       " 'GAB3': [],\n",
       " 'TOR2A': [],\n",
       " 'DLL3': [],\n",
       " 'DLGAP3': [],\n",
       " 'PRKG1': [],\n",
       " 'CRH': [],\n",
       " 'SAA1': [],\n",
       " 'SIX3': []}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possibletargets = { i : j for i,j in gene2path.items() if len(j) > 1 }\n",
    "possibletargets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Those are our 2 targets. Let's look at the data over time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Genes, Cells) = (0, 221)\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py:3361\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3359'>3360</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3360'>3361</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3361'>3362</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/_libs/index.pyx:76\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/_libs/index.pyx:108\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5198\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5206\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Gene'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb Cell 30'\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=4'>5</a>\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(folder\u001b[39m+\u001b[39mf,delimiter\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m'\u001b[39m)[ [i \u001b[39min\u001b[39;00m possibletargets \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m df[\u001b[39m'\u001b[39m\u001b[39mGene\u001b[39m\u001b[39m'\u001b[39m]] ]\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=5'>6</a>\u001b[0m counts \u001b[39m=\u001b[39m df[df\u001b[39m.\u001b[39mcolumns[\u001b[39m1\u001b[39m:]]\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=6'>7</a>\u001b[0m genes \u001b[39m=\u001b[39m df[\u001b[39m'\u001b[39;49m\u001b[39mGene\u001b[39;49m\u001b[39m'\u001b[39;49m]\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=7'>8</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39m(Genes, Cells) =\u001b[39m\u001b[39m'\u001b[39m,counts\u001b[39m.\u001b[39mshape)\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/ryan/Documents/Interview/RNAseqAnalysis/ProcessingV0.ipynb#ch0000030?line=8'>9</a>\u001b[0m genemean \u001b[39m=\u001b[39m counts\u001b[39m.\u001b[39mmean(axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py:3458\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3455'>3456</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mnlevels \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3456'>3457</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3457'>3458</a>\u001b[0m indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(key)\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3458'>3459</a>\u001b[0m \u001b[39mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/frame.py?line=3459'>3460</a>\u001b[0m     indexer \u001b[39m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py:3363\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3360'>3361</a>\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3361'>3362</a>\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m-> <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3362'>3363</a>\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3364'>3365</a>\u001b[0m \u001b[39mif\u001b[39;00m is_scalar(key) \u001b[39mand\u001b[39;00m isna(key) \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhasnans:\n\u001b[1;32m   <a href='file:///~/anaconda3/envs/chemprop/lib/python3.9/site-packages/pandas/core/indexes/base.py?line=3365'>3366</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Gene'"
     ]
    }
   ],
   "source": [
    "folder = 'DenoisedData/'\n",
    "stds = []\n",
    "means = []\n",
    "for i,f in enumerate(sorted(listdir(folder))):\n",
    "    df = pd.read_csv(folder+f,delimiter='\\t')[ [i in possibletargets for i in df['Gene']] ]\n",
    "    counts = df[df.columns[1:]]\n",
    "    genes = df['Gene']\n",
    "    print('(Genes, Cells) =',counts.shape)\n",
    "    genemean = counts.mean(axis=1)\n",
    "    genestd = counts.std(axis=1)\n",
    "\n",
    "    \n",
    "    stds += [genestd]\n",
    "    means += [genemean]\n",
    "\n",
    "stds = np.array(list(stds))\n",
    "means = np.array(list(means))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.363897847707054 44.275400568066885\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXIAAAD5CAYAAAA6JL6mAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAP+ElEQVR4nO3df2yd1X3H8c+HOEChWWHEMEbIzDo0rUAbJI+2oqgdo1s2UGGTKujataompUhbSauOQpmWUdROItNQxdRpRC0FqRk/WmBDJdMKaQmLxAAHQsEERGgJhCTYARLMWhLH/u6Pew1O7OAb3+P73O+975dk5d7Hl3PPCc7H33ue85zHESEAQF6HVd0BAEBzCHIASI4gB4DkCHIASI4gB4DkCHIASK6nVEO250kakPRSRFzwTq9duHBh9PX1lXprAOgKGzZs2BkRvQceLxbkkpZL2iTp12Z6YV9fnwYGBgq+NQB0PttbpjteZGrF9iJJ50v6Ton2AACNKzVH/i1JX5U0Xqg9AECDmg5y2xdIGoqIDTO8bpntAdsDw8PDzb4tAKCuREV+tqRP2H5e0q2SzrX9/QNfFBGrIqI/Ivp7e6fM1QMAZqnpII+Ir0XEoojok3SJpJ9ExGea7hkAoCGsIweA5EouP1RE3C/p/pJtAgDeGRU5ACRHkAPd7Hvn176QGkEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACqx71Dm0KQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAtMocXcFKkANAck0Hue0jbT9s+3Hbg7a/XqJjAIDG9BRoY4+kcyPiDdvzJa23/V8R8b8F2gYAzKDpII+IkPRG/en8+lc02y4AoDFF5shtz7O9UdKQpHsj4qES7QJAJxncvluD23cXb7dIkEfEWEQskbRI0lm2Tz/wNbaX2R6wPTA8PFzibQEAKrxqJSJ2Sbpf0tJpvrcqIvojor+3t7fk2wJAVyuxaqXX9jH1x++SdJ6kp5ttF0D3mKsph25RYtXKiZJutj1PtV8Mt0fEjwq0CwBoQIlVKz+TdGaBvgAAZoErOwEgOYIcAJIjyAEguRInOwEkNbFS5LSK+4HmUJEDQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5ADQHIEOQAkR5DP4OIbHtTFNzxYdTcA4KAIcgBIjiAHgOQIcgBIjiAHgOQIcgBIjiAHgOQI8gnfO7/2BQDJEOQAkBxBDgDJEeQAkBxBjkqw9QFQDkEOAMkR5ACQHEEOAMkR5ACQHEEOAMk1HeS2T7b9U9ubbA/aXl6iYwCAxvQUaGOfpK9ExKO2F0jaYPveiHiqQNsAgBk0XZFHxPaIeLT+eETSJkknNdsuAKAxRefIbfdJOlPSQyXbBQAcXLEgt/1uSXdI+lJEvD7N95fZHrA9MDw8XOpt0WG44hM4dEWC3PZ81UJ8dUTcOd1rImJVRPRHRH9vb2+JtwUAqMyqFUv6rqRNEXFd810CAByKEqtWzpb0l5KesL2xfuyqiFhToG10mRWvXF5/tL7SfgCZNB3kEbFekgv0BQAwC6mu7OREGABMlSrIAQBTEeQAkBxBDgDJEeQAkBxBDqByPTGqvtHnpJGXq+5KSgQ5gMotHBvSUfFLad21VXclJYIc6GJtUQmP7NCx46/VLkbZuJqqfBYIcqCLtUUlvG6lpKg9jnGq8lkgyIFu1Q6V8MgOaePqt4NobG9HV+UjY/M1/uYb2rnjhaLtEuRAt2qHSnjdytp7T9bBVfme0TGd5ue1+QcrirZLkAPdqF0q4a0P1957srG9teMdZue2Lfqgn9JhDi3ZeU/RqpwgB7pRu1TCl66Xrt6twcPP0ODhZ0hX7659Xdp5u18+d8cKuf4JyBovWpUT5EA36qJKuB3s3LZFH9h5j47wPknSEd5XtCovsR85gGzqFe/gP35EknTaVZ1XAbeT5+5YoSUT5yPqJqryhV+8qen2qcgBYI4t3PWzt6rxCUd4n3p3PV6kfSpyAJV7dfxoXfuri3TjyJs6fsGRVXenuPf+/WOSpn4Cem+h9qnIAVTulj1na3Bska5fu7nqrqREkAOo1NDrb+q+0fcrdJh+OPCihkberLpL6RDkACp1/dpnNV6/7e9YBFX5LBDkACoz9Pqb+sGGrdpXP103OhZU5bNAkAOozPVrn9V47L8sj6r80BHkACrz6Au7NDq2f5CPjoUe3fJaRT3KieWHACqzZvk5krgwqVkEOSqx4pXL64/4hws0i6kVAEiOIAe62KvjR+uK//s0q0SSI8iBLsYVlZ2BIAe6FFdUdg6CHOhSXFHZOQhyoAtxRWVnIcjrBrfv1uD23VOO7903rqe2vc4PODoKV1R2FoJ8Bi/t+pVG9uzjBxwdhSsqOwsXBNX1xKgW7XuhdhfxBSdIqn38HH5jjyTphwMv6rI//J2O3PQe3YcrKjtLkYrc9o22h2w/WaK9KiwcG9JR8cv97iJ+/dpnNXGbPT52AmhXpaZWbpK0tFBbrTeyQ8eOv1Y7f79xtTTy8lsngyY+fHIyqDV6YlR9o8/VPhkBaEiRII+IByS9WqKtSqxbqbdK7xiX1l3LyaCKTPfJCMA7a9nJTtvLbA/YHhgeHp5VGyteuXzSZkuFjOyQNq5++y9ibK+0cbV+8fwvOBnUaiM7dEz9k1HUPxkBmFnLTnZGxCpJqySpv78/Znh566xbWavCJ4txrT71funL1+niGx6UJN32hQ+3vm/dZt1Kjcdhmucxje0bU8+6a6ULrqu6V0DbY/nh1odrVfhkY3trxzFnpmzWNLJD8dhqzfeYJKlHo1TlQINYfnhpbdnVwZZhsW/23Ji8WdM3LjpdWrdSY2Nj+/1AUpUDjSm1/PAWSQ9K+l3bW23/VYl20Zmm26xpdMtD6tHofq/r0ahGtzxUUS+BPIpU5BHxqRLtoDtMu1nTb/6bbtv24n4nmOfPsy4+abG+UVVHgSSYWkFLHWyzpsXHHcUqIWCWCHK01MHW5591ynH68Zc/yiXjwCywagUtxWZNQHlU5GgpNmsCyqMiB4DkCHIASI4gB4DkCHIASI4gB4DkCHJUghtIAOUQ5KgEN5AAykkV5FO2PkVO09xaD8DspQryyVufIrFpbq0HYPbSBPl0W58ioYPcWo+qHJi9NEE+7danyOcgt9ajKgdmL0WQH2zrU6ryhLi1HlBciiCf2Pq0V6/ptsOvUa92tawqZ5lcYZeul67ercHDz9Dg4WdIV++ufV3K5lnAbKUI8omtTy/ruUu/72f0xZ47W7b1KcvkALS7FNvYrll+jjSyQ+P//GkdptBnj1yvzy7/V2nBCXP7xgcuk/voFXP/ngBwiFJU5JLmfMnatGvUWSYHIIEcQd6CJWtT1qizTA5AEjmCfI6XrE27Rp1lcgCSyBHkc7xkbdo16iyTA5BEipOdE0vT5uI+jwdbo37ZFffp+AVHcm9JAG0vR0U+hybWqE/GlaMAMun6IJ9Yoz5Zq9aoA0AJOaZW5tCa5edImptpGwBoha6vyAEgu66vyAGgVU478T1z0i4VOQAkR5ADQHIEOQAkR5ADQHJFgtz2UtvP2N5s+8oSbQIAGtP0qhXb8yR9W9LHJW2V9IjtuyPiqWbbBoCO8vl75qTZEhX5WZI2R8TPI2KvpFslXVigXQBAA0qsIz9J0ouTnm+V9MEDX2R7maRlkrR48eICb4tOdM1x/yRJuq3ifgCZlAhyT3MsphyIWCVplST19/dP+T6A7sUv8OaUmFrZKunkSc8XSdpWoF0AQANKVOSPSDrV9imSXpJ0iaS/KNAugC5x2xc+XHUXUms6yCNin+2/kfTfkuZJujEiBpvuGQCgIUU2zYqINZLWlGir3TB3B6DdsfshKnGwX5B8xAYOHZfoA0ByqSpypjkAYCoqcgBIjiAHgOQIcgBILtUc+Vxi/h1AVlTkAJAcQQ4AyTG1MgMuUAHQ7qjIASA5ghwAkmNqBehip534nqq7gAKoyAEgOSryOk5qthZ/30A5VOQAkBxBDgDJMbUCdLPP31N1D1AAFTkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByBDkAJEeQA0ByqTbNYg9rAJiKihwAkiPIASC5poLc9idtD9oet91fqlMAgMY1W5E/KenPJT1QoC8AgFlo6mRnRGySJNtlegMAOGQtmyO3vcz2gO2B4eHhVr0tAHS8GSty2/dJ+o1pvvV3EfGfjb5RRKyStEqS+vv7o+EeAgDe0YxBHhHntaIjAIDZYfkhACTniNnPctj+M0n/IqlX0i5JGyPijxv474YlbZnhZQsl7Zx15/LohnF2wxil7hhnN4xRat9x/lZE9B54sKkgn0u2ByKi49emd8M4u2GMUneMsxvGKOUbJ1MrAJAcQQ4AybVzkK+qugMt0g3j7IYxSt0xzm4Yo5RsnG07Rw4AaEw7V+QAgAa0ZZDbXmr7GdubbV9ZdX9KsX2j7SHbT0469uu277X9bP3PY6vsY7Nsn2z7p7Y31XfGXF4/3jHjtH2k7YdtP14f49frxztmjBNsz7P9mO0f1Z934hift/2E7Y22B+rHUo2z7YLc9jxJ35b0J5LeJ+lTtt9Xba+KuUnS0gOOXSlpbUScKmlt/Xlm+yR9JSJ+T9KHJP11/f9fJ41zj6RzI+IDkpZIWmr7Q+qsMU5YLmnTpOedOEZJ+oOIWDJpyWGqcbZdkEs6S9LmiPh5ROyVdKukCyvuUxER8YCkVw84fKGkm+uPb5Z0USv7VFpEbI+IR+uPR1QLgZPUQeOMmjfqT+fXv0IdNEZJsr1I0vmSvjPpcEeN8R2kGmc7BvlJkl6c9Hxr/VinOiEitku1EJR0fMX9KcZ2n6QzJT2kDhtnfcpho6QhSfdGRMeNUdK3JH1V0vikY502Rqn2S/jHtjfYXlY/lmqc7Xjz5ek2N2dpTTK23y3pDklfiojXO23P+ogYk7TE9jGS7rJ9esVdKsr2BZKGImKD7Y9V3J25dnZEbLN9vKR7bT9ddYcOVTtW5FslnTzp+SJJ2yrqSyu8bPtESar/OVRxf5pme75qIb46Iu6sH+64cUpSROySdL9q5z46aYxnS/qE7edVm9481/b31VljlCRFxLb6n0OS7lJtejfVONsxyB+RdKrtU2wfLukSSXdX3Ke5dLekz9Uff05Sw3u8tyPXSu/vStoUEddN+lbHjNN2b70Sl+13STpP0tPqoDFGxNciYlFE9Kn2b/AnEfEZddAYJcn20bYXTDyW9Eeq3cIy1Tjb8oIg23+q2vzcPEk3RsQ3q+1RGbZvkfQx1XZWe1nSP0j6D0m3S1os6QVJn4yIA0+IpmH7I5L+R9ITentu9SrV5sk7Ypy236/aCbB5qhVDt0fENbaPU4eMcbL61MrfRsQFnTZG27+tWhUu1aaa/z0ivpltnG0Z5ACAxrXj1AoA4BAQ5ACQHEEOAMkR5ACQHEEOAMkR5ACQHEEOAMkR5ACQ3P8DvoFTovhHOmoAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "t = [int(i[:2]) for i in sorted(listdir(folder))]\n",
    "plt.errorbar(t, means[:,0]/np.median(means[:,0]), stds[:,0]/np.median(means[:,0]), linestyle='None', marker='^')\n",
    "plt.errorbar(t, means[:,1]/np.median(means[:,1]), stds[:,1]/np.median(means[:,1]), linestyle='None', marker='^')\n",
    "print(np.std(means[:,1]),max(stds[:,1]))\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "569d0bbc3854db4a73f5be0bb5576848176d13611564839ca8e34057b1299597"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
